// @faker-js/faker/dist/cjs/modules/internet/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e8) { throw _e8; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e9) { didErr = true; err = _e9; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var w = Object.create;
var k = Object.defineProperty;
var C = Object.getOwnPropertyDescriptor;
var $ = Object.getOwnPropertyNames;
var y = Object.getPrototypeOf,
  x = Object.prototype.hasOwnProperty;
var v = function v(a, e) {
    for (var r in e) k(a, r, {
      get: e[r],
      enumerable: !0
    });
  },
  S = function S(a, e, r, t) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper($(e)),
        _step;
      try {
        var _loop = function _loop() {
          var n = _step.value;
          !x.call(a, n) && n !== r && k(a, n, {
            get: function get() {
              return e[n];
            },
            enumerable: !(t = C(e, n)) || t.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return a;
  };
var j = function j(a, e, r) {
    return r = a != null ? w(y(a)) : {}, S(e || !a || !a.__esModule ? k(r, "default", {
      value: a,
      enumerable: !0
    }) : r, a);
  },
  _ = function _(a) {
    return S(k({}, "__esModule", {
      value: !0
    }), a);
  };
var B = {};
v(B, {
  InternetModule: function InternetModule() {
    return E;
  }
});
module.exports = _(B);
var p = require("../../internal/deprecated"),
  N = require("./char-mappings"),
  A = j(require("./user-agent"));
var E = /*#__PURE__*/function () {
  function E(e) {
    _classCallCheck(this, E);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(E.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var r = _step2.value;
        r === "constructor" || typeof this[r] != "function" || (this[r] = this[r].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(E, [{
    key: "avatar",
    value: function avatar() {
      return "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/".concat(this.faker.number["int"](1249), ".jpg");
    }
  }, {
    key: "email",
    value: function email() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      var t = arguments.length > 2 ? arguments[2] : undefined;
      var n = arguments.length > 3 ? arguments[3] : undefined;
      var f;
      (typeof e == "string" || r != null || t != null || n != null) && (0, p.deprecated)({
        deprecated: "faker.internet.email(firstName, lastName, provider, options)",
        proposed: "faker.internet.email({ firstName, lastName, provider, ... })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "string" && (e = {
        firstName: e
      });
      var _e = e,
        _e$firstName = _e.firstName,
        s = _e$firstName === void 0 ? this.faker.person.firstName() : _e$firstName,
        _e$lastName = _e.lastName,
        o = _e$lastName === void 0 ? r != null ? r : this.faker.person.lastName() : _e$lastName,
        _e$provider = _e.provider,
        l = _e$provider === void 0 ? t != null ? t : this.faker.helpers.arrayElement(this.faker.definitions.internet.free_email) : _e$provider,
        _e$allowSpecialCharac = _e.allowSpecialCharacters,
        m = _e$allowSpecialCharac === void 0 ? (f = n == null ? void 0 : n.allowSpecialCharacters) != null ? f : !1 : _e$allowSpecialCharac;
      var i = this.userName({
        firstName: s,
        lastName: o
      });
      if (i = i.replace(/[^A-Za-z0-9._+\-]+/g, ""), i = i.substring(0, 50), m) {
        var u = "._-".split(""),
          h = ".!#$%&'*+-/=?^_`{|}~".split("");
        i = i.replace(this.faker.helpers.arrayElement(u), this.faker.helpers.arrayElement(h));
      }
      return i = i.replace(/\.{2,}/g, "."), i = i.replace(/^\./, ""), i = i.replace(/\.$/, ""), "".concat(i, "@").concat(l);
    }
  }, {
    key: "exampleEmail",
    value: function exampleEmail() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      var t = arguments.length > 2 ? arguments[2] : undefined;
      var m;
      (typeof e == "string" || r != null || t != null) && (0, p.deprecated)({
        deprecated: "faker.internet.exampleEmail(firstName, lastName, options)",
        proposed: "faker.internet.exampleEmail({ firstName, lastName, ... })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "string" && (e = {
        firstName: e
      });
      var _e2 = e,
        _e2$firstName = _e2.firstName,
        n = _e2$firstName === void 0 ? this.faker.person.firstName() : _e2$firstName,
        _e2$lastName = _e2.lastName,
        s = _e2$lastName === void 0 ? r != null ? r : this.faker.person.lastName() : _e2$lastName,
        _e2$allowSpecialChara = _e2.allowSpecialCharacters,
        o = _e2$allowSpecialChara === void 0 ? (m = t == null ? void 0 : t.allowSpecialCharacters) != null ? m : !1 : _e2$allowSpecialChara,
        l = this.faker.helpers.arrayElement(this.faker.definitions.internet.example_email);
      return this.email({
        firstName: n,
        lastName: s,
        provider: l,
        allowSpecialCharacters: o
      });
    }
  }, {
    key: "userName",
    value: function userName() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      (typeof e == "string" || r != null) && (0, p.deprecated)({
        deprecated: "faker.internet.userName(firstName, lastName)",
        proposed: "faker.internet.userName({ firstName, lastName })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "string" && (e = {
        firstName: e
      });
      var _e3 = e,
        _e3$firstName = _e3.firstName,
        t = _e3$firstName === void 0 ? this.faker.person.firstName() : _e3$firstName,
        _e3$lastName = _e3.lastName,
        n = _e3$lastName === void 0 ? r != null ? r : this.faker.person.lastName() : _e3$lastName;
      var s;
      switch (this.faker.number["int"](2)) {
        case 0:
          s = "".concat(t).concat(this.faker.number["int"](99));
          break;
        case 1:
          s = t + this.faker.helpers.arrayElement([".", "_"]) + n;
          break;
        case 2:
          s = "".concat(t).concat(this.faker.helpers.arrayElement([".", "_"])).concat(n).concat(this.faker.number["int"](99));
          break;
      }
      return s = s.normalize("NFKD").replace(/[\u0300-\u036f]/g, ""), s = s.split("").map(function (o) {
        return N.charMapping[o] ? N.charMapping[o] : o.charCodeAt(0) < 128 ? o : o.charCodeAt(0).toString(36);
      }).join(""), s = s.toString().replace(/'/g, ""), s = s.replace(/ /g, ""), s;
    }
  }, {
    key: "displayName",
    value: function displayName() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      (typeof e == "string" || r != null) && (0, p.deprecated)({
        deprecated: "faker.internet.displayName(firstName, lastName)",
        proposed: "faker.internet.displayName({ firstName, lastName })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "string" && (e = {
        firstName: e
      });
      var _e4 = e,
        _e4$firstName = _e4.firstName,
        t = _e4$firstName === void 0 ? this.faker.person.firstName() : _e4$firstName,
        _e4$lastName = _e4.lastName,
        n = _e4$lastName === void 0 ? r != null ? r : this.faker.person.lastName() : _e4$lastName;
      var s;
      switch (this.faker.number["int"](2)) {
        case 0:
          s = "".concat(t).concat(this.faker.number["int"](99));
          break;
        case 1:
          s = t + this.faker.helpers.arrayElement([".", "_"]) + n;
          break;
        case 2:
          s = "".concat(t).concat(this.faker.helpers.arrayElement([".", "_"])).concat(n).concat(this.faker.number["int"](99));
          break;
      }
      return s = s.toString().replace(/'/g, ""), s = s.replace(/ /g, ""), s;
    }
  }, {
    key: "protocol",
    value: function protocol() {
      var e = ["http", "https"];
      return this.faker.helpers.arrayElement(e);
    }
  }, {
    key: "httpMethod",
    value: function httpMethod() {
      var e = ["GET", "POST", "PUT", "DELETE", "PATCH"];
      return this.faker.helpers.arrayElement(e);
    }
  }, {
    key: "httpStatusCode",
    value: function httpStatusCode() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$types = e.types,
        r = _e$types === void 0 ? Object.keys(this.faker.definitions.internet.http_status_code) : _e$types,
        t = this.faker.helpers.arrayElement(r);
      return this.faker.helpers.arrayElement(this.faker.definitions.internet.http_status_code[t]);
    }
  }, {
    key: "url",
    value: function url() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$appendSlash = e.appendSlash,
        r = _e$appendSlash === void 0 ? this.faker.datatype["boolean"]() : _e$appendSlash,
        _e$protocol = e.protocol,
        t = _e$protocol === void 0 ? "https" : _e$protocol;
      return "".concat(t, "://").concat(this.domainName()).concat(r ? "/" : "");
    }
  }, {
    key: "domainName",
    value: function domainName() {
      return "".concat(this.domainWord(), ".").concat(this.domainSuffix());
    }
  }, {
    key: "domainSuffix",
    value: function domainSuffix() {
      return this.faker.helpers.arrayElement(this.faker.definitions.internet.domain_suffix);
    }
  }, {
    key: "domainWord",
    value: function domainWord() {
      return this.faker.helpers.slugify("".concat(this.faker.word.adjective(), "-").concat(this.faker.word.noun())).toLowerCase();
    }
  }, {
    key: "ip",
    value: function ip() {
      return this.faker.datatype["boolean"]() ? this.ipv4() : this.ipv6();
    }
  }, {
    key: "ipv4",
    value: function ipv4() {
      var _this = this;
      return Array.from({
        length: 4
      }, function () {
        return _this.faker.number["int"](255);
      }).join(".");
    }
  }, {
    key: "ipv6",
    value: function ipv6() {
      var _this2 = this;
      return Array.from({
        length: 8
      }, function () {
        return _this2.faker.string.hexadecimal({
          length: 4,
          casing: "lower",
          prefix: ""
        });
      }).join(":");
    }
  }, {
    key: "port",
    value: function port() {
      return this.faker.number["int"](65535);
    }
  }, {
    key: "userAgent",
    value: function userAgent() {
      return A.generate(this.faker);
    }
  }, {
    key: "color",
    value: function color() {
      var _this3 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      var t = arguments.length > 2 ? arguments[2] : undefined;
      (typeof e == "number" || t != null || r != null) && (0, p.deprecated)({
        deprecated: "faker.internet.color(redBase, greenBase, blueBase)",
        proposed: "faker.internet.color({ redBase, greenBase, blueBase })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "number" && (e = {
        redBase: e
      });
      var _e5 = e,
        _e5$redBase = _e5.redBase,
        n = _e5$redBase === void 0 ? 0 : _e5$redBase,
        _e5$greenBase = _e5.greenBase,
        s = _e5$greenBase === void 0 ? r != null ? r : 0 : _e5$greenBase,
        _e5$blueBase = _e5.blueBase,
        o = _e5$blueBase === void 0 ? t != null ? t : 0 : _e5$blueBase,
        l = function l(u) {
          return Math.floor((_this3.faker.number["int"](256) + u) / 2).toString(16).padStart(2, "0");
        },
        m = l(n),
        i = l(s),
        f = l(o);
      return "#".concat(m).concat(i).concat(f);
    }
  }, {
    key: "mac",
    value: function mac() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      typeof e == "string" && (e = {
        separator: e
      });
      var _e6 = e,
        _e6$separator = _e6.separator,
        r = _e6$separator === void 0 ? ":" : _e6$separator,
        t,
        n = "";
      for ([":", "-", ""].includes(r) || (r = ":"), t = 0; t < 12; t++) n += this.faker.number.hex(15), t % 2 === 1 && t !== 11 && (n += r);
      return n;
    }
  }, {
    key: "password",
    value: function password() {
      var _this4 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 ? arguments[1] : undefined;
      var t = arguments.length > 2 ? arguments[2] : undefined;
      var n = arguments.length > 3 ? arguments[3] : undefined;
      var s = /[aeiouAEIOU]$/,
        o = /[bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ]$/,
        l = function l(h, c, g, d) {
          if (d.length >= h) return d;
          c && (o.test(d) ? g = s : g = o);
          var T = _this4.faker.number["int"](94) + 33;
          var b = String.fromCharCode(T);
          return c && (b = b.toLowerCase()), g.test(b) ? l(h, c, g, d + b) : l(h, c, g, d);
        };
      (typeof e == "string" || r != null || t != null || n != null) && (0, p.deprecated)({
        deprecated: "faker.internet.password(length, memorable, pattern, prefix)",
        proposed: "faker.internet.password({ length, memorable, pattern, prefix })",
        since: "8.0",
        until: "9.0"
      }), typeof e == "number" && (e = {
        length: e
      });
      var _e7 = e,
        _e7$length = _e7.length,
        m = _e7$length === void 0 ? 15 : _e7$length,
        _e7$memorable = _e7.memorable,
        i = _e7$memorable === void 0 ? r != null ? r : !1 : _e7$memorable,
        _e7$pattern = _e7.pattern,
        f = _e7$pattern === void 0 ? t != null ? t : /\w/ : _e7$pattern,
        _e7$prefix = _e7.prefix,
        u = _e7$prefix === void 0 ? n != null ? n : "" : _e7$prefix;
      return l(m, i, f, u);
    }
  }, {
    key: "emoji",
    value: function emoji() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$types2 = e.types,
        r = _e$types2 === void 0 ? Object.keys(this.faker.definitions.internet.emoji) : _e$types2,
        t = this.faker.helpers.arrayElement(r);
      return this.faker.helpers.arrayElement(this.faker.definitions.internet.emoji[t]);
    }
  }]);
  return E;
}();
0 && (module.exports = {
  InternetModule: InternetModule
});
