// @faker-js/faker/dist/cjs/modules/internet/user-agent.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArrayLimit(arr, i) { var _i = null == arr ? null : "undefined" != typeof Symbol && arr[Symbol.iterator] || arr["@@iterator"]; if (null != _i) { var _s, _e, _x, _r, _arr = [], _n = !0, _d = !1; try { if (_x = (_i = _i.call(arr)).next, 0 === i) { if (Object(_i) !== _i) return; _n = !1; } else for (; !(_n = (_s = _x.call(_i)).done) && (_arr.push(_s.value), _arr.length !== i); _n = !0); } catch (err) { _d = !0, _e = err; } finally { try { if (!_n && null != _i["return"] && (_r = _i["return"](), Object(_r) !== _r)) return; } finally { if (_d) throw _e; } } return _arr; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var u = Object.defineProperty;
var x = Object.getOwnPropertyDescriptor;
var p = Object.getOwnPropertyNames;
var S = Object.prototype.hasOwnProperty;
var O = function O(n, t) {
    for (var $ in t) u(n, $, {
      get: t[$],
      enumerable: !0
    });
  },
  w = function w(n, t, $, s) {
    if (t && _typeof(t) == "object" || typeof t == "function") {
      var _iterator = _createForOfIteratorHelper(p(t)),
        _step;
      try {
        var _loop = function _loop() {
          var m = _step.value;
          !S.call(n, m) && m !== $ && u(n, m, {
            get: function get() {
              return t[m];
            },
            enumerable: !(s = x(t, m)) || s.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return n;
  };
var M = function M(n) {
  return w(u({}, "__esModule", {
    value: !0
  }), n);
};
var T = {};
O(T, {
  generate: function generate() {
    return d;
  }
});
module.exports = M(T);
function d(n) {
  var t = function t() {
      return n.helpers.arrayElement(["AB", "AF", "AN", "AR", "AS", "AZ", "BE", "BG", "BN", "BO", "BR", "BS", "CA", "CE", "CO", "CS", "CU", "CY", "DA", "DE", "EL", "EN", "EO", "ES", "ET", "EU", "FA", "FI", "FJ", "FO", "FR", "FY", "GA", "GD", "GL", "GV", "HE", "HI", "HR", "HT", "HU", "HY", "ID", "IS", "IT", "JA", "JV", "KA", "KG", "KO", "KU", "KW", "KY", "LA", "LB", "LI", "LN", "LT", "LV", "MG", "MK", "MN", "MO", "MS", "MT", "MY", "NB", "NE", "NL", "NN", "NO", "OC", "PL", "PT", "RM", "RO", "RU", "SC", "SE", "SK", "SL", "SO", "SQ", "SR", "SV", "SW", "TK", "TR", "TY", "UK", "UR", "UZ", "VI", "VO", "YI", "ZH"]);
    },
    $ = function $() {
      var r = {
          chrome: ["win", "mac", "lin"],
          firefox: ["win", "mac", "lin"],
          opera: ["win", "mac", "lin"],
          safari: ["win", "mac"],
          iexplorer: ["win"]
        },
        e = n.helpers.objectKey(r),
        o = n.helpers.arrayElement(r[e]);
      return [e, o];
    },
    s = function s(r) {
      return n.helpers.arrayElement({
        lin: ["i686", "x86_64"],
        mac: ["Intel", "PPC", "U; Intel", "U; PPC"],
        win: ["", "WOW64", "Win64; x64"]
      }[r]);
    },
    m = function m(r) {
      var e = "";
      for (var o = 0; o < r; o++) e += ".".concat(n.string.numeric({
        allowLeadingZeros: !0
      }));
      return e;
    },
    i = {
      net: function net() {
        return [n.number["int"]({
          min: 1,
          max: 4
        }), n.number["int"](9), n.number["int"]({
          min: 1e4,
          max: 99999
        }), n.number["int"](9)].join(".");
      },
      nt: function nt() {
        return [n.number["int"]({
          min: 5,
          max: 6
        }), n.number["int"](3)].join(".");
      },
      ie: function ie() {
        return n.number["int"]({
          min: 7,
          max: 11
        });
      },
      trident: function trident() {
        return [n.number["int"]({
          min: 3,
          max: 7
        }), n.number["int"](1)].join(".");
      },
      osx: function osx(r) {
        return [10, n.number["int"]({
          min: 5,
          max: 10
        }), n.number["int"](9)].join(r || ".");
      },
      chrome: function chrome() {
        return [n.number["int"]({
          min: 13,
          max: 39
        }), 0, n.number["int"]({
          min: 800,
          max: 899
        }), 0].join(".");
      },
      presto: function presto() {
        return "2.9.".concat(n.number["int"]({
          min: 160,
          max: 190
        }));
      },
      presto2: function presto2() {
        return "".concat(n.number["int"]({
          min: 10,
          max: 12
        }), ".00");
      },
      safari: function safari() {
        return [n.number["int"]({
          min: 531,
          max: 538
        }), n.number["int"](2), n.number["int"](2)].join(".");
      }
    },
    c = {
      firefox: function firefox(r) {
        var e = "".concat(n.number["int"]({
            min: 5,
            max: 15
          })).concat(m(2)),
          o = "Gecko/20100101 Firefox/".concat(e),
          a = s(r);
        return "Mozilla/5.0 ".concat(r === "win" ? "(Windows NT ".concat(i.nt()).concat(a ? "; ".concat(a) : "") : r === "mac" ? "(Macintosh; ".concat(a, " Mac OS X ").concat(i.osx()) : "(X11; Linux ".concat(a), "; rv:").concat(e.slice(0, -2), ") ").concat(o);
      },
      iexplorer: function iexplorer() {
        var r = i.ie();
        return r >= 11 ? "Mozilla/5.0 (Windows NT 6.".concat(n.number["int"]({
          min: 1,
          max: 3
        }), "; Trident/7.0; ").concat(n.datatype["boolean"]() ? "Touch; " : "", "rv:11.0) like Gecko") : "Mozilla/5.0 (compatible; MSIE ".concat(r, ".0; Windows NT ").concat(i.nt(), "; Trident/").concat(i.trident()).concat(n.datatype["boolean"]() ? "; .NET CLR ".concat(i.net()) : "", ")");
      },
      opera: function opera(r) {
        var e = " Presto/".concat(i.presto(), " Version/").concat(i.presto2(), ")"),
          o = r === "win" ? "(Windows NT ".concat(i.nt(), "; U; ").concat(t()).concat(e) : r === "lin" ? "(X11; Linux ".concat(s(r), "; U; ").concat(t()).concat(e) : "(Macintosh; Intel Mac OS X ".concat(i.osx(), " U; ").concat(t(), " Presto/").concat(i.presto(), " Version/").concat(i.presto2(), ")");
        return "Opera/".concat(n.number["int"]({
          min: 9,
          max: 14
        }), ".").concat(n.number["int"](99), " ").concat(o);
      },
      safari: function safari(r) {
        var e = i.safari(),
          o = "".concat(n.number["int"]({
            min: 4,
            max: 7
          }), ".").concat(n.number["int"](1), ".").concat(n.number["int"](10));
        return "Mozilla/5.0 ".concat(r === "mac" ? "(Macintosh; ".concat(s("mac"), " Mac OS X ").concat(i.osx("_"), " rv:").concat(n.number["int"]({
          min: 2,
          max: 6
        }), ".0; ").concat(t(), ") ") : "(Windows; U; Windows NT ".concat(i.nt(), ")"), "AppleWebKit/").concat(e, " (KHTML, like Gecko) Version/").concat(o, " Safari/").concat(e);
      },
      chrome: function chrome(r) {
        var e = i.safari();
        return "Mozilla/5.0 ".concat(r === "mac" ? "(Macintosh; ".concat(s("mac"), " Mac OS X ").concat(i.osx("_"), ") ") : r === "win" ? "(Windows; U; Windows NT ".concat(i.nt(), ")") : "(X11; Linux ".concat(s(r)), " AppleWebKit/").concat(e, " (KHTML, like Gecko) Chrome/").concat(i.chrome(), " Safari/").concat(e);
      }
    },
    _$ = $(),
    _$2 = _slicedToArray(_$, 2),
    l = _$2[0],
    b = _$2[1];
  return c[l](b);
}
0 && (module.exports = {
  generate: generate
});
