// @faker-js/faker/dist/cjs/modules/airline/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var n = Object.defineProperty;
var p = Object.getOwnPropertyDescriptor;
var d = Object.getOwnPropertyNames;
var f = Object.prototype.hasOwnProperty;
var y = function y(t, r) {
    for (var e in r) n(t, e, {
      get: r[e],
      enumerable: !0
    });
  },
  u = function u(t, r, e, i) {
    if (r && _typeof(r) == "object" || typeof r == "function") {
      var _iterator = _createForOfIteratorHelper(d(r)),
        _step;
      try {
        var _loop = function _loop() {
          var a = _step.value;
          !f.call(t, a) && a !== e && n(t, a, {
            get: function get() {
              return r[a];
            },
            enumerable: !(i = p(r, a)) || i.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return t;
  };
var m = function m(t) {
  return u(n({}, "__esModule", {
    value: !0
  }), t);
};
var k = {};
y(k, {
  Aircraft: function Aircraft() {
    return o;
  },
  AirlineModule: function AirlineModule() {
    return s;
  }
});
module.exports = m(k);
var o = function (i) {
  return i.Narrowbody = "narrowbody", i.Regional = "regional", i.Widebody = "widebody", i;
}(o || {});
var h = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"],
  g = ["0", "O", "1", "I", "L"],
  b = {
    regional: 20,
    narrowbody: 35,
    widebody: 60
  },
  w = {
    regional: ["A", "B", "C", "D"],
    narrowbody: ["A", "B", "C", "D", "E", "F"],
    widebody: ["A", "B", "C", "D", "E", "F", "G", "H", "J", "K"]
  };
var s = /*#__PURE__*/function () {
  function s(r) {
    _classCallCheck(this, s);
    this.faker = r;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(s.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var e = _step2.value;
        e === "constructor" || typeof this[e] != "function" || (this[e] = this[e].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(s, [{
    key: "airport",
    value: function airport() {
      return this.faker.helpers.arrayElement(this.faker.definitions.airline.airport);
    }
  }, {
    key: "airline",
    value: function airline() {
      return this.faker.helpers.arrayElement(this.faker.definitions.airline.airline);
    }
  }, {
    key: "airplane",
    value: function airplane() {
      return this.faker.helpers.arrayElement(this.faker.definitions.airline.airplane);
    }
  }, {
    key: "recordLocator",
    value: function recordLocator() {
      var r = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _r$allowNumerics = r.allowNumerics,
        e = _r$allowNumerics === void 0 ? !1 : _r$allowNumerics,
        _r$allowVisuallySimil = r.allowVisuallySimilarCharacters,
        i = _r$allowVisuallySimil === void 0 ? !1 : _r$allowVisuallySimil,
        a = [];
      return e || a.push.apply(a, h), i || a.push.apply(a, g), this.faker.string.alphanumeric({
        length: 6,
        casing: "upper",
        exclude: a
      });
    }
  }, {
    key: "seat",
    value: function seat() {
      var r = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _r$aircraftType = r.aircraftType,
        e = _r$aircraftType === void 0 ? "narrowbody" : _r$aircraftType,
        i = b[e],
        a = w[e],
        l = this.faker.number["int"]({
          min: 1,
          max: i
        }),
        c = this.faker.helpers.arrayElement(a);
      return "".concat(l).concat(c);
    }
  }, {
    key: "aircraftType",
    value: function aircraftType() {
      return this.faker.helpers.enumValue(o);
    }
  }, {
    key: "flightNumber",
    value: function flightNumber() {
      var r = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _r$length = r.length,
        e = _r$length === void 0 ? {
          min: 1,
          max: 4
        } : _r$length,
        _r$addLeadingZeros = r.addLeadingZeros,
        i = _r$addLeadingZeros === void 0 ? !1 : _r$addLeadingZeros,
        a = this.faker.string.numeric({
          length: e,
          allowLeadingZeros: !1
        });
      return i ? a.padStart(4, "0") : a;
    }
  }]);
  return s;
}();
0 && (module.exports = {
  Aircraft: Aircraft,
  AirlineModule: AirlineModule
});
