// @redis/bloom/dist/commands/cuckoo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pushInsertOptions = void 0;
var ADD = require("./ADD");
var ADDNX = require("./ADDNX");
var COUNT = require("./COUNT");
var DEL = require("./DEL");
var EXISTS = require("./EXISTS");
var INFO = require("./INFO");
var INSERT = require("./INSERT");
var INSERTNX = require("./INSERTNX");
var LOADCHUNK = require("./LOADCHUNK");
var RESERVE = require("./RESERVE");
var SCANDUMP = require("./SCANDUMP");
var generic_transformers_1 = require("@redis/client/dist/lib/commands/generic-transformers");
exports["default"] = {
  ADD: ADD,
  add: ADD,
  ADDNX: ADDNX,
  addNX: ADDNX,
  COUNT: COUNT,
  count: COUNT,
  DEL: DEL,
  del: DEL,
  EXISTS: EXISTS,
  exists: EXISTS,
  INFO: INFO,
  info: INFO,
  INSERT: INSERT,
  insert: INSERT,
  INSERTNX: INSERTNX,
  insertNX: INSERTNX,
  LOADCHUNK: LOADCHUNK,
  loadChunk: LOADCHUNK,
  RESERVE: RESERVE,
  reserve: RESERVE,
  SCANDUMP: SCANDUMP,
  scanDump: SCANDUMP
};
function pushInsertOptions(args, items, options) {
  if (options !== null && options !== void 0 && options.CAPACITY) {
    args.push('CAPACITY');
    args.push(options.CAPACITY.toString());
  }
  if (options !== null && options !== void 0 && options.NOCREATE) {
    args.push('NOCREATE');
  }
  args.push('ITEMS');
  return (0, generic_transformers_1.pushVerdictArguments)(args, items);
}
exports.pushInsertOptions = pushInsertOptions;
