// @faker-js/faker/dist/cjs/modules/company/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var s = Object.defineProperty;
var c = Object.getOwnPropertyDescriptor;
var p = Object.getOwnPropertyNames;
var u = Object.prototype.hasOwnProperty;
var f = function f(t, e) {
    for (var r in e) s(t, r, {
      get: e[r],
      enumerable: !0
    });
  },
  h = function h(t, e, r, o) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(p(e)),
        _step;
      try {
        var _loop = function _loop() {
          var i = _step.value;
          !u.call(t, i) && i !== r && s(t, i, {
            get: function get() {
              return e[i];
            },
            enumerable: !(o = c(e, i)) || o.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return t;
  };
var d = function d(t) {
  return h(s({}, "__esModule", {
    value: !0
  }), t);
};
var m = {};
f(m, {
  CompanyModule: function CompanyModule() {
    return a;
  }
});
module.exports = d(m);
var n = require("../../internal/deprecated");
var a = /*#__PURE__*/function () {
  function a(e) {
    _classCallCheck(this, a);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(a.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var r = _step2.value;
        r === "constructor" || typeof this[r] != "function" || (this[r] = this[r].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(a, [{
    key: "suffixes",
    value: function suffixes() {
      return (0, n.deprecated)({
        deprecated: "faker.company.suffixes",
        proposed: "faker.company.name",
        since: "8.0",
        until: "9.0"
      }), this.faker.definitions.company.suffix.slice(0);
    }
  }, {
    key: "name",
    value: function name() {
      return this.faker.helpers.fake(this.faker.definitions.company.name_pattern);
    }
  }, {
    key: "companySuffix",
    value: function companySuffix() {
      return (0, n.deprecated)({
        deprecated: "faker.company.companySuffix",
        proposed: "faker.company.name",
        since: "8.0",
        until: "9.0"
      }), this.faker.helpers.arrayElement(this.suffixes());
    }
  }, {
    key: "catchPhrase",
    value: function catchPhrase() {
      return [this.catchPhraseAdjective(), this.catchPhraseDescriptor(), this.catchPhraseNoun()].join(" ");
    }
  }, {
    key: "bs",
    value: function bs() {
      return (0, n.deprecated)({
        deprecated: "faker.company.bs",
        proposed: "faker.company.buzzPhrase",
        since: "8.0",
        until: "9.0"
      }), this.buzzPhrase();
    }
  }, {
    key: "buzzPhrase",
    value: function buzzPhrase() {
      return [this.buzzVerb(), this.buzzAdjective(), this.buzzNoun()].join(" ");
    }
  }, {
    key: "catchPhraseAdjective",
    value: function catchPhraseAdjective() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.adjective);
    }
  }, {
    key: "catchPhraseDescriptor",
    value: function catchPhraseDescriptor() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.descriptor);
    }
  }, {
    key: "catchPhraseNoun",
    value: function catchPhraseNoun() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.noun);
    }
  }, {
    key: "bsAdjective",
    value: function bsAdjective() {
      return (0, n.deprecated)({
        deprecated: "faker.company.bsAdjective",
        proposed: "faker.company.buzzAdjective",
        since: "8.0",
        until: "9.0"
      }), this.buzzAdjective();
    }
  }, {
    key: "buzzAdjective",
    value: function buzzAdjective() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.buzz_adjective);
    }
  }, {
    key: "bsBuzz",
    value: function bsBuzz() {
      return (0, n.deprecated)({
        deprecated: "faker.company.bsBuzz",
        proposed: "faker.company.buzzVerb",
        since: "8.0",
        until: "9.0"
      }), this.buzzVerb();
    }
  }, {
    key: "buzzVerb",
    value: function buzzVerb() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.buzz_verb);
    }
  }, {
    key: "bsNoun",
    value: function bsNoun() {
      return (0, n.deprecated)({
        deprecated: "faker.company.bsNoun",
        proposed: "faker.company.buzzNoun",
        since: "8.0",
        until: "9.0"
      }), this.buzzNoun();
    }
  }, {
    key: "buzzNoun",
    value: function buzzNoun() {
      return this.faker.helpers.arrayElement(this.faker.definitions.company.buzz_noun);
    }
  }]);
  return a;
}();
0 && (module.exports = {
  CompanyModule: CompanyModule
});
