// @faker-js/faker/dist/cjs/modules/commerce/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var m = Object.defineProperty;
var b = Object.getOwnPropertyDescriptor;
var f = Object.getOwnPropertyNames;
var l = Object.prototype.hasOwnProperty;
var y = function y(t, e) {
    for (var r in e) m(t, r, {
      get: e[r],
      enumerable: !0
    });
  },
  h = function h(t, e, r, i) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(f(e)),
        _step;
      try {
        var _loop = function _loop() {
          var n = _step.value;
          !l.call(t, n) && n !== r && m(t, n, {
            get: function get() {
              return e[n];
            },
            enumerable: !(i = b(e, n)) || i.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return t;
  };
var g = function g(t) {
  return h(m({}, "__esModule", {
    value: !0
  }), t);
};
var k = {};
y(k, {
  CommerceModule: function CommerceModule() {
    return c;
  }
});
module.exports = g(k);
var u = require("../../internal/deprecated");
var c = /*#__PURE__*/function () {
  function c(e) {
    _classCallCheck(this, c);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(c.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var r = _step2.value;
        r === "constructor" || typeof this[r] != "function" || (this[r] = this[r].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(c, [{
    key: "department",
    value: function department() {
      return this.faker.helpers.arrayElement(this.faker.definitions.commerce.department);
    }
  }, {
    key: "productName",
    value: function productName() {
      return "".concat(this.productAdjective(), " ").concat(this.productMaterial(), " ").concat(this.product());
    }
  }, {
    key: "price",
    value: function price() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1e3;
      var i = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 2;
      var n = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "";
      typeof e == "number" && ((0, u.deprecated)({
        deprecated: "faker.commerce.price(min, max, dec, symbol)",
        proposed: "faker.commerce.price({ min, max, dec, symbol })",
        since: "8.0",
        until: "9.0"
      }), e = {
        min: e,
        dec: i,
        max: r,
        symbol: n
      });
      var _e = e,
        _e$dec = _e.dec,
        d = _e$dec === void 0 ? 2 : _e$dec,
        _e$max = _e.max,
        s = _e$max === void 0 ? 1e3 : _e$max,
        _e$min = _e.min,
        o = _e$min === void 0 ? 1 : _e$min,
        _e$symbol = _e.symbol,
        a = _e$symbol === void 0 ? "" : _e$symbol;
      if (o < 0 || s < 0) return "".concat(a, "0");
      var p = this.faker.number["int"]({
        min: o,
        max: s
      });
      return a + p.toFixed(d);
    }
  }, {
    key: "productAdjective",
    value: function productAdjective() {
      return this.faker.helpers.arrayElement(this.faker.definitions.commerce.product_name.adjective);
    }
  }, {
    key: "productMaterial",
    value: function productMaterial() {
      return this.faker.helpers.arrayElement(this.faker.definitions.commerce.product_name.material);
    }
  }, {
    key: "product",
    value: function product() {
      return this.faker.helpers.arrayElement(this.faker.definitions.commerce.product_name.product);
    }
  }, {
    key: "productDescription",
    value: function productDescription() {
      return this.faker.helpers.arrayElement(this.faker.definitions.commerce.product_description);
    }
  }]);
  return c;
}();
0 && (module.exports = {
  CommerceModule: CommerceModule
});
