// dayjs/esm/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Y = exports.W = exports.SECONDS_A_WEEK = exports.SECONDS_A_MINUTE = exports.SECONDS_A_HOUR = exports.SECONDS_A_DAY = exports.S = exports.REGEX_PARSE = exports.REGEX_FORMAT = exports.Q = exports.MS = exports.MIN = exports.MILLISECONDS_A_WEEK = exports.MILLISECONDS_A_SECOND = exports.MILLISECONDS_A_MINUTE = exports.MILLISECONDS_A_HOUR = exports.MILLISECONDS_A_DAY = exports.M = exports.INVALID_DATE_STRING = exports.H = exports.FORMAT_DEFAULT = exports.DATE = exports.D = void 0;
var SECONDS_A_MINUTE = 60;
exports.SECONDS_A_MINUTE = SECONDS_A_MINUTE;
var SECONDS_A_HOUR = SECONDS_A_MINUTE * 60;
exports.SECONDS_A_HOUR = SECONDS_A_HOUR;
var SECONDS_A_DAY = SECONDS_A_HOUR * 24;
exports.SECONDS_A_DAY = SECONDS_A_DAY;
var SECONDS_A_WEEK = SECONDS_A_DAY * 7;
exports.SECONDS_A_WEEK = SECONDS_A_WEEK;
var MILLISECONDS_A_SECOND = 1e3;
exports.MILLISECONDS_A_SECOND = MILLISECONDS_A_SECOND;
var MILLISECONDS_A_MINUTE = SECONDS_A_MINUTE * MILLISECONDS_A_SECOND;
exports.MILLISECONDS_A_MINUTE = MILLISECONDS_A_MINUTE;
var MILLISECONDS_A_HOUR = SECONDS_A_HOUR * MILLISECONDS_A_SECOND;
exports.MILLISECONDS_A_HOUR = MILLISECONDS_A_HOUR;
var MILLISECONDS_A_DAY = SECONDS_A_DAY * MILLISECONDS_A_SECOND;
exports.MILLISECONDS_A_DAY = MILLISECONDS_A_DAY;
var MILLISECONDS_A_WEEK = SECONDS_A_WEEK * MILLISECONDS_A_SECOND; // English locales
exports.MILLISECONDS_A_WEEK = MILLISECONDS_A_WEEK;
var MS = 'millisecond';
exports.MS = MS;
var S = 'second';
exports.S = S;
var MIN = 'minute';
exports.MIN = MIN;
var H = 'hour';
exports.H = H;
var D = 'day';
exports.D = D;
var W = 'week';
exports.W = W;
var M = 'month';
exports.M = M;
var Q = 'quarter';
exports.Q = Q;
var Y = 'year';
exports.Y = Y;
var DATE = 'date';
exports.DATE = DATE;
var FORMAT_DEFAULT = 'YYYY-MM-DDTHH:mm:ssZ';
exports.FORMAT_DEFAULT = FORMAT_DEFAULT;
var INVALID_DATE_STRING = 'Invalid Date'; // regex
exports.INVALID_DATE_STRING = INVALID_DATE_STRING;
var REGEX_PARSE = /^(\d{4})[-/]?(\d{1,2})?[-/]?(\d{0,2})[Tt\s]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?[.:]?(\d+)?$/;
exports.REGEX_PARSE = REGEX_PARSE;
var REGEX_FORMAT = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g;
exports.REGEX_FORMAT = REGEX_FORMAT;
