// @faker-js/faker/dist/cjs/faker.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e4) { throw _e4; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e5) { didErr = true; err = _e5; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var V = Object.create;
var a = Object.defineProperty;
var W = Object.getOwnPropertyDescriptor;
var X = Object.getOwnPropertyNames;
var j = Object.getPrototypeOf,
  q = Object.prototype.hasOwnProperty;
var z = function z(o, e) {
    for (var r in e) a(o, r, {
      get: e[r],
      enumerable: !0
    });
  },
  s = function s(o, e, r, n) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(X(e)),
        _step;
      try {
        var _loop = function _loop() {
          var l = _step.value;
          !q.call(o, l) && l !== r && a(o, l, {
            get: function get() {
              return e[l];
            },
            enumerable: !(n = W(e, l)) || n.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return o;
  };
var B = function B(o, e, r) {
    return r = o != null ? V(j(o)) : {}, s(e || !o || !o.__esModule ? a(r, "default", {
      value: o,
      enumerable: !0
    }) : r, o);
  },
  J = function J(o) {
    return s(a({}, "__esModule", {
      value: !0
    }), o);
  };
var Q = {};
z(Q, {
  Faker: function Faker() {
    return K;
  }
});
module.exports = J(Q);
var t = require("./errors/faker-error"),
  i = require("./internal/deprecated"),
  d = B(require("./internal/mersenne/mersenne")),
  c = require("./locale-proxy"),
  m = require("./modules/airline"),
  u = require("./modules/animal"),
  p = require("./modules/color"),
  f = require("./modules/commerce"),
  h = require("./modules/company"),
  M = require("./modules/database"),
  y = require("./modules/datatype"),
  w = require("./modules/date"),
  b = require("./modules/finance"),
  D = require("./modules/git"),
  g = require("./modules/hacker"),
  v = require("./modules/helpers"),
  k = require("./modules/image"),
  L = require("./modules/internet"),
  F = require("./modules/location"),
  P = require("./modules/lorem"),
  R = require("./modules/music"),
  A = require("./modules/number"),
  T = require("./modules/person"),
  _ = require("./modules/phone"),
  C = require("./modules/random"),
  x = require("./modules/science"),
  N = require("./modules/string"),
  I = require("./modules/system"),
  S = require("./modules/vehicle"),
  E = require("./modules/word"),
  H = require("./utils/merge-locales");
var K = /*#__PURE__*/function () {
  function K(e) {
    _classCallCheck(this, K);
    this._defaultRefDate = function () {
      return new Date();
    };
    this._mersenne = (0, d["default"])();
    this.random = new C.RandomModule(this);
    this.helpers = new v.HelpersModule(this);
    this.datatype = new y.DatatypeModule(this);
    this.airline = new m.AirlineModule(this);
    this.animal = new u.AnimalModule(this);
    this.color = new p.ColorModule(this);
    this.commerce = new f.CommerceModule(this);
    this.company = new h.CompanyModule(this);
    this.database = new M.DatabaseModule(this);
    this.date = new w.DateModule(this);
    this.finance = new b.FinanceModule(this);
    this.git = new D.GitModule(this);
    this.hacker = new g.HackerModule(this);
    this.image = new k.ImageModule(this);
    this.internet = new L.InternetModule(this);
    this.location = new F.LocationModule(this);
    this.lorem = new P.LoremModule(this);
    this.music = new R.MusicModule(this);
    this.person = new T.PersonModule(this);
    this.number = new A.NumberModule(this);
    this.phone = new _.PhoneModule(this);
    this.science = new x.ScienceModule(this);
    this.string = new N.StringModule(this);
    this.system = new I.SystemModule(this);
    this.vehicle = new S.VehicleModule(this);
    this.word = new E.WordModule(this);
    var _e = e,
      r = _e.locales;
    if (r != null) {
      (0, i.deprecated)({
        deprecated: "new Faker({ locales: {a, b}, locale: 'a', localeFallback: 'b' })",
        proposed: "new Faker({ locale: [a, b, ...] }) or new Faker({ locale: a })",
        since: "8.0",
        until: "9.0"
      });
      var _e2 = e,
        _e2$locale = _e2.locale,
        l = _e2$locale === void 0 ? "en" : _e2$locale,
        _e2$localeFallback = _e2.localeFallback,
        G = _e2$localeFallback === void 0 ? "en" : _e2$localeFallback;
      e = {
        locale: [r[l], r[G]]
      };
    }
    var _e3 = e,
      n = _e3.locale;
    if (Array.isArray(n)) {
      if (n.length === 0) throw new t.FakerError("The locale option must contain at least one locale definition.");
      n = (0, H.mergeLocales)(n);
    }
    this.rawDefinitions = n, this.definitions = (0, c.createLocaleProxy)(this.rawDefinitions);
  }
  _createClass(K, [{
    key: "defaultRefDate",
    get: function get() {
      return this._defaultRefDate;
    }
  }, {
    key: "setDefaultRefDate",
    value: function setDefaultRefDate() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : function () {
        return new Date();
      };
      typeof e == "function" ? this._defaultRefDate = e : this._defaultRefDate = function () {
        return new Date(e);
      };
    }
  }, {
    key: "address",
    get: function get() {
      return (0, i.deprecated)({
        deprecated: "faker.address",
        proposed: "faker.location",
        since: "8.0",
        until: "10.0"
      }), this.location;
    }
  }, {
    key: "name",
    get: function get() {
      return (0, i.deprecated)({
        deprecated: "faker.name",
        proposed: "faker.person",
        since: "8.0",
        until: "10.0"
      }), this.person;
    }
  }, {
    key: "seed",
    value: function seed() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Math.ceil(Math.random() * Number.MAX_SAFE_INTEGER);
      return this._mersenne.seed(e), e;
    }
  }, {
    key: "locales",
    get: function get() {
      throw new t.FakerError("The locales property has been removed. Please use the constructor instead.");
    },
    set: function set(e) {
      throw new t.FakerError("The locales property has been removed. Please use the constructor instead.");
    }
  }, {
    key: "locale",
    get: function get() {
      throw new t.FakerError("The locale property has been removed. Please use the constructor instead.");
    },
    set: function set(e) {
      throw new t.FakerError("The locale property has been removed. Please use the constructor instead.");
    }
  }, {
    key: "localeFallback",
    get: function get() {
      throw new t.FakerError("The localeFallback property has been removed. Please use the constructor instead.");
    },
    set: function set(e) {
      throw new t.FakerError("The localeFallback property has been removed. Please use the constructor instead.");
    }
  }, {
    key: "setLocale",
    value: function setLocale() {
      throw new t.FakerError("This method has been removed. Please use the constructor instead.");
    }
  }]);
  return K;
}();
0 && (module.exports = {
  Faker: Faker
});
