// @faker-js/faker/dist/cjs/modules/system/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var l = Object.defineProperty;
var x = Object.getOwnPropertyDescriptor;
var $ = Object.getOwnPropertyNames;
var T = Object.prototype.hasOwnProperty;
var S = function S(a, e) {
    for (var t in e) l(a, t, {
      get: e[t],
      enumerable: !0
    });
  },
  w = function w(a, e, t, s) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper($(e)),
        _step;
      try {
        var _loop = function _loop() {
          var r = _step.value;
          !T.call(a, r) && r !== t && l(a, r, {
            get: function get() {
              return e[r];
            },
            enumerable: !(s = x(e, r)) || s.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return a;
  };
var N = function N(a) {
  return w(l({}, "__esModule", {
    value: !0
  }), a);
};
var A = {};
S(A, {
  SystemModule: function SystemModule() {
    return y;
  }
});
module.exports = N(A);
var O = ["video", "audio", "image", "text", "application"],
  F = ["application/pdf", "audio/mpeg", "audio/wav", "image/png", "image/jpeg", "image/gif", "video/mp4", "video/mpeg", "text/html"],
  j = ["en", "wl", "ww"],
  k = {
    index: "o",
    slot: "s",
    mac: "x",
    pci: "p"
  },
  v = ["SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT"];
var y = /*#__PURE__*/function () {
  function y(e) {
    _classCallCheck(this, y);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(y.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var t = _step2.value;
        t === "constructor" || typeof this[t] != "function" || (this[t] = this[t].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(y, [{
    key: "fileName",
    value: function fileName() {
      var _this = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$extensionCount = e.extensionCount,
        t = _e$extensionCount === void 0 ? 1 : _e$extensionCount,
        s = this.faker.word.words().toLowerCase().replace(/\W/g, "_"),
        r = this.faker.helpers.multiple(function () {
          return _this.fileExt();
        }, {
          count: t
        }).join(".");
      return r.length === 0 ? s : "".concat(s, ".").concat(r);
    }
  }, {
    key: "commonFileName",
    value: function commonFileName(e) {
      return "".concat(this.fileName({
        extensionCount: 0
      }), ".").concat(e || this.commonFileExt());
    }
  }, {
    key: "mimeType",
    value: function mimeType() {
      var e = Object.keys(this.faker.definitions.system.mimeTypes);
      return this.faker.helpers.arrayElement(e);
    }
  }, {
    key: "commonFileType",
    value: function commonFileType() {
      return this.faker.helpers.arrayElement(O);
    }
  }, {
    key: "commonFileExt",
    value: function commonFileExt() {
      return this.fileExt(this.faker.helpers.arrayElement(F));
    }
  }, {
    key: "fileType",
    value: function fileType() {
      var e = new Set(),
        t = this.faker.definitions.system.mimeTypes;
      Object.keys(t).forEach(function (r) {
        var i = r.split("/")[0];
        e.add(i);
      });
      var s = Array.from(e);
      return this.faker.helpers.arrayElement(s);
    }
  }, {
    key: "fileExt",
    value: function fileExt(e) {
      if (typeof e == "string") {
        var i = this.faker.definitions.system.mimeTypes;
        return this.faker.helpers.arrayElement(i[e].extensions);
      }
      var t = this.faker.definitions.system.mimeTypes,
        s = new Set();
      Object.keys(t).forEach(function (i) {
        t[i].extensions instanceof Array && t[i].extensions.forEach(function (n) {
          s.add(n);
        });
      });
      var r = Array.from(s);
      return this.faker.helpers.arrayElement(r);
    }
  }, {
    key: "directoryPath",
    value: function directoryPath() {
      var e = this.faker.definitions.system.directoryPaths;
      return this.faker.helpers.arrayElement(e);
    }
  }, {
    key: "filePath",
    value: function filePath() {
      return "".concat(this.directoryPath(), "/").concat(this.fileName());
    }
  }, {
    key: "semver",
    value: function semver() {
      return [this.faker.number["int"](9), this.faker.number["int"](9), this.faker.number["int"](9)].join(".");
    }
  }, {
    key: "networkInterface",
    value: function networkInterface() {
      var _this2 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var o, m, c, f, h;
      var _e$interfaceType = e.interfaceType,
        t = _e$interfaceType === void 0 ? this.faker.helpers.arrayElement(j) : _e$interfaceType,
        _e$interfaceSchema = e.interfaceSchema,
        s = _e$interfaceSchema === void 0 ? this.faker.helpers.objectKey(k) : _e$interfaceSchema;
      var r,
        i = "";
      var n = function n() {
        return _this2.faker.string.numeric({
          allowLeadingZeros: !0
        });
      };
      switch (s) {
        case "index":
          r = n();
          break;
        case "slot":
          r = "".concat(n()).concat((o = this.faker.helpers.maybe(function () {
            return "f".concat(n());
          })) != null ? o : "").concat((m = this.faker.helpers.maybe(function () {
            return "d".concat(n());
          })) != null ? m : "");
          break;
        case "mac":
          r = this.faker.internet.mac("");
          break;
        case "pci":
          i = (c = this.faker.helpers.maybe(function () {
            return "P".concat(n());
          })) != null ? c : "", r = "".concat(n(), "s").concat(n()).concat((f = this.faker.helpers.maybe(function () {
            return "f".concat(n());
          })) != null ? f : "").concat((h = this.faker.helpers.maybe(function () {
            return "d".concat(n());
          })) != null ? h : "");
          break;
      }
      return "".concat(i).concat(t).concat(k[s]).concat(r);
    }
  }, {
    key: "cron",
    value: function cron() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$includeYear = e.includeYear,
        t = _e$includeYear === void 0 ? !1 : _e$includeYear,
        _e$includeNonStandard = e.includeNonStandard,
        s = _e$includeNonStandard === void 0 ? !1 : _e$includeNonStandard,
        r = [this.faker.number["int"](59), "*"],
        i = [this.faker.number["int"](23), "*"],
        n = [this.faker.number["int"]({
          min: 1,
          max: 31
        }), "*", "?"],
        o = [this.faker.number["int"]({
          min: 1,
          max: 12
        }), "*"],
        m = [this.faker.number["int"](6), this.faker.helpers.arrayElement(v), "*", "?"],
        c = [this.faker.number["int"]({
          min: 1970,
          max: 2099
        }), "*"],
        f = this.faker.helpers.arrayElement(r),
        h = this.faker.helpers.arrayElement(i),
        u = this.faker.helpers.arrayElement(n),
        d = this.faker.helpers.arrayElement(o),
        b = this.faker.helpers.arrayElement(m),
        g = this.faker.helpers.arrayElement(c);
      var p = "".concat(f, " ").concat(h, " ").concat(u, " ").concat(d, " ").concat(b);
      t && (p += " ".concat(g));
      var E = ["@annually", "@daily", "@hourly", "@monthly", "@reboot", "@weekly", "@yearly"];
      return !s || this.faker.datatype["boolean"]() ? p : this.faker.helpers.arrayElement(E);
    }
  }]);
  return y;
}();
0 && (module.exports = {
  SystemModule: SystemModule
});
