// cypress/lib/errors.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var chalk = require('chalk');
var _require = require('common-tags'),
  stripIndent = _require.stripIndent,
  stripIndents = _require.stripIndents;
var la = require('lazy-ass');
var is = require('check-more-types');
var util = require('./util');
var state = require('./tasks/state');
var docsUrl = 'https://on.cypress.io';
var requiredDependenciesUrl = "".concat(docsUrl, "/required-dependencies");
var runDocumentationUrl = "".concat(docsUrl, "/cypress-run");

// TODO it would be nice if all error objects could be enforced via types
// to only have description + solution properties

var hr = '----------';
var genericErrorSolution = stripIndent(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  Search for an existing issue or open a GitHub issue at\n\n    ", "\n"])), chalk.blue(util.issuesUrl));

// common errors Cypress application can encounter
var unknownError = {
  description: 'Unknown Cypress CLI error',
  solution: genericErrorSolution
};
var invalidRunProjectPath = {
  description: 'Invalid --project path',
  solution: stripIndent(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    Please provide a valid project path.\n\n    Learn more about ", " at:\n\n      ", "\n  "])), chalk.cyan('cypress run'), chalk.blue(runDocumentationUrl))
};
var invalidOS = {
  description: 'The Cypress App could not be installed. Your machine does not meet the operating system requirements.',
  solution: stripIndent(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\n  ", ""])), chalk.blue('https://on.cypress.io/guides/getting-started/installing-cypress#system-requirements'))
};
var failedDownload = {
  description: 'The Cypress App could not be downloaded.',
  solution: stripIndent(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration\n\n  Otherwise, please check network connectivity and try again:"])))
};
var failedUnzip = {
  description: 'The Cypress App could not be unzipped.',
  solution: genericErrorSolution
};
var failedUnzipWindowsMaxPathLength = {
  description: 'The Cypress App could not be unzipped.',
  solution: "This is most likely because the maximum path length is being exceeded on your system.\n\n  Read here for solutions to this problem: https://on.cypress.io/win-max-path-length-error"
};
var missingApp = function missingApp(binaryDir) {
  return {
    description: "No version of Cypress is installed in: ".concat(chalk.cyan(binaryDir)),
    solution: stripIndent(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n    \nPlease reinstall Cypress by running: ", "\n  "], ["\n    \\nPlease reinstall Cypress by running: ", "\n  "])), chalk.cyan('cypress install'))
  };
};
var binaryNotExecutable = function binaryNotExecutable(executable) {
  return {
    description: "Cypress cannot run because this binary file does not have executable permissions here:\n\n".concat(executable),
    solution: stripIndent(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n\n    Reasons this may happen:\n\n    - node was installed as 'root' or with 'sudo'\n    - the cypress npm package as 'root' or with 'sudo'\n\n    Please check that you have the appropriate user permissions.\n\n    You can also try clearing the cache with 'cypress cache clear' and reinstalling.\n  "], ["\\n\n    Reasons this may happen:\n\n    - node was installed as 'root' or with 'sudo'\n    - the cypress npm package as 'root' or with 'sudo'\n\n    Please check that you have the appropriate user permissions.\n\n    You can also try clearing the cache with 'cypress cache clear' and reinstalling.\n  "])))
  };
};
var notInstalledCI = function notInstalledCI(executable) {
  return {
    description: 'The cypress npm package is installed, but the Cypress binary is missing.',
    solution: stripIndent(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n\n    We expected the binary to be installed here: ", "\n\n    Reasons it may be missing:\n\n    - You're caching 'node_modules' but are not caching this path: ", "\n    - You ran 'npm install' at an earlier build step but did not persist: ", "\n\n    Properly caching the binary will fix this error and avoid downloading and unzipping Cypress.\n\n    Alternatively, you can run 'cypress install' to download the binary again.\n\n    ", "\n  "], ["\\n\n    We expected the binary to be installed here: ", "\n\n    Reasons it may be missing:\n\n    - You're caching 'node_modules' but are not caching this path: ", "\n    - You ran 'npm install' at an earlier build step but did not persist: ", "\n\n    Properly caching the binary will fix this error and avoid downloading and unzipping Cypress.\n\n    Alternatively, you can run 'cypress install' to download the binary again.\n\n    ", "\n  "])), chalk.cyan(executable), util.getCacheDir(), util.getCacheDir(), chalk.blue('https://on.cypress.io/not-installed-ci-error'))
  };
};
var nonZeroExitCodeXvfb = {
  description: 'Xvfb exited with a non zero exit code.',
  solution: stripIndent(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["\n    There was a problem spawning Xvfb.\n\n    This is likely a problem with your system, permissions, or installation of Xvfb.\n    "])))
};
var missingXvfb = {
  description: 'Your system is missing the dependency: Xvfb',
  solution: stripIndent(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["\n    Install Xvfb and run Cypress again.\n\n    Read our documentation on dependencies for more information:\n\n      ", "\n\n    If you are using Docker, we provide containers with all required dependencies installed.\n    "])), chalk.blue(requiredDependenciesUrl))
};
var smokeTestFailure = function smokeTestFailure(smokeTestCommand, timedOut) {
  return {
    description: "Cypress verification ".concat(timedOut ? 'timed out' : 'failed', "."),
    solution: stripIndent(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["\n    This command failed with the following output:\n\n    ", "\n\n    "])), smokeTestCommand)
  };
};
var invalidSmokeTestDisplayError = {
  code: 'INVALID_SMOKE_TEST_DISPLAY_ERROR',
  description: 'Cypress verification failed.',
  solution: function solution(msg) {
    return stripIndent(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["\n      Cypress failed to start after spawning a new Xvfb server.\n\n      The error logs we received were:\n\n      ", "\n\n      ", "\n\n      ", "\n\n      This may be due to a missing library or dependency. ", "\n\n      Please refer to the error above for more detail.\n    "])), hr, msg, hr, chalk.blue(requiredDependenciesUrl));
  }
};
var missingDependency = {
  description: 'Cypress failed to start.',
  // this message is too Linux specific
  solution: stripIndent(_templateObject12 || (_templateObject12 = _taggedTemplateLiteral(["\n    This may be due to a missing library or dependency. ", "\n\n    Please refer to the error below for more details.\n  "])), chalk.blue(requiredDependenciesUrl))
};
var invalidCacheDirectory = {
  description: 'Cypress cannot write to the cache directory due to file permissions',
  solution: stripIndent(_templateObject13 || (_templateObject13 = _taggedTemplateLiteral(["\n    See discussion and possible solutions at\n    ", "\n  "])), chalk.blue(util.getGitHubIssueUrl(1281)))
};
var versionMismatch = {
  description: 'Installed version does not match package version.',
  solution: 'Install Cypress and verify app again'
};
var incompatibleHeadlessFlags = {
  description: '`--headed` and `--headless` cannot both be passed.',
  solution: 'Either pass `--headed` or `--headless`, but not both.'
};
var solutionUnknown = stripIndent(_templateObject14 || (_templateObject14 = _taggedTemplateLiteral(["\n  Please search Cypress documentation for possible solutions:\n\n    ", "\n\n  Check if there is a GitHub issue describing this crash:\n\n    ", "\n\n  Consider opening a new issue.\n"])), chalk.blue(docsUrl), chalk.blue(util.issuesUrl));
var unexpected = {
  description: 'An unexpected error occurred while verifying the Cypress executable.',
  solution: solutionUnknown
};
var invalidCypressEnv = {
  description: chalk.red('The environment variable with the reserved name "CYPRESS_INTERNAL_ENV" is set.'),
  solution: chalk.red('Unset the "CYPRESS_INTERNAL_ENV" environment variable and run Cypress again.'),
  exitCode: 11
};
var invalidTestingType = {
  description: 'Invalid testingType',
  solution: "Please provide a valid testingType. Valid test types are ".concat(chalk.cyan('\'e2e\''), " and ").concat(chalk.cyan('\'component\''), ".")
};
var incompatibleTestTypeFlags = {
  description: '`--e2e` and `--component` cannot both be passed.',
  solution: 'Either pass `--e2e` or `--component`, but not both.'
};
var incompatibleTestingTypeAndFlag = {
  description: 'Set a `testingType` and also passed `--e2e` or `--component` flags.',
  solution: 'Either set `testingType` or pass a testing type flag, but not both.'
};
var invalidConfigFile = {
  description: '`--config-file` cannot be false.',
  solution: 'Either pass a relative path to a valid Cypress config file or remove this option.'
};

/**
 * This error happens when CLI detects that the child Test Runner process
 * was killed with a signal, like SIGBUS
 * @see https://github.com/cypress-io/cypress/issues/5808
 * @param {'close'|'event'} eventName Child close event name
 * @param {string} signal Signal that closed the child process, like "SIGBUS"
*/
var childProcessKilled = function childProcessKilled(eventName, signal) {
  return {
    description: "The Test Runner unexpectedly exited via a ".concat(chalk.cyan(eventName), " event with signal ").concat(chalk.cyan(signal)),
    solution: solutionUnknown
  };
};
var CYPRESS_RUN_BINARY = {
  notValid: function notValid(value) {
    var properFormat = "**/".concat(state.getPlatformExecutable());
    return {
      description: "Could not run binary set by environment variable: CYPRESS_RUN_BINARY=".concat(value),
      solution: "Ensure the environment variable is a path to the Cypress binary, matching ".concat(properFormat)
    };
  }
};
function addPlatformInformation(info) {
  return util.getPlatformInfo().then(function (platform) {
    return _objectSpread(_objectSpread({}, info), {}, {
      platform: platform
    });
  });
}

/**
 * Given an error object (see the errors above), forms error message text with details,
 * then resolves with Error instance you can throw or reject with.
 * @param {object} errorObject
 * @returns {Promise<Error>} resolves with an Error
 * @example
  ```js
  // inside a Promise with "resolve" and "reject"
  const errorObject = childProcessKilled('exit', 'SIGKILL')
  return getError(errorObject).then(reject)
  ```
 */
function getError(errorObject) {
  return formErrorText(errorObject).then(function (errorMessage) {
    var err = new Error(errorMessage);
    err.known = true;
    return err;
  });
}

/**
 * Forms nice error message with error and platform information,
 * and if possible a way to solve it. Resolves with a string.
 */
function formErrorText(info, msg, prevMessage) {
  return addPlatformInformation(info).then(function (obj) {
    var formatted = [];
    function add(msg) {
      formatted.push(stripIndents(msg));
    }
    la(is.unemptyString(obj.description), 'expected error description to be text', obj.description);

    // assuming that if there the solution is a function it will handle
    // error message and (optional previous error message)
    if (is.fn(obj.solution)) {
      var text = obj.solution(msg, prevMessage);
      la(is.unemptyString(text), 'expected solution to be text', text);
      add("\n        ".concat(obj.description, "\n\n        ").concat(text, "\n\n      "));
    } else {
      la(is.unemptyString(obj.solution), 'expected error solution to be text', obj.solution);
      add("\n        ".concat(obj.description, "\n\n        ").concat(obj.solution, "\n\n      "));
      if (msg) {
        add("\n          ".concat(hr, "\n\n          ").concat(msg, "\n\n        "));
      }
    }
    add("\n      ".concat(hr, "\n\n      ").concat(obj.platform, "\n    "));
    if (obj.footer) {
      add("\n\n        ".concat(hr, "\n\n        ").concat(obj.footer, "\n      "));
    }
    return formatted.join('\n\n');
  });
}
var raise = function raise(info) {
  return function (text) {
    var err = new Error(text);
    if (info.code) {
      err.code = info.code;
    }
    err.known = true;
    throw err;
  };
};
var throwFormErrorText = function throwFormErrorText(info) {
  return function (msg, prevMessage) {
    return formErrorText(info, msg, prevMessage).then(raise(info));
  };
};

/**
 * Forms full error message with error and OS details, prints to the error output
 * and then exits the process.
 * @param {ErrorInformation} info Error information {description, solution}
 * @example return exitWithError(errors.invalidCypressEnv)('foo')
 */
var exitWithError = function exitWithError(info) {
  return function (msg) {
    return formErrorText(info, msg).then(function (text) {
      // eslint-disable-next-line no-console
      console.error(text);
      process.exit(info.exitCode || 1);
    });
  };
};
module.exports = {
  raise: raise,
  exitWithError: exitWithError,
  // formError,
  formErrorText: formErrorText,
  throwFormErrorText: throwFormErrorText,
  getError: getError,
  hr: hr,
  errors: {
    unknownError: unknownError,
    nonZeroExitCodeXvfb: nonZeroExitCodeXvfb,
    missingXvfb: missingXvfb,
    missingApp: missingApp,
    notInstalledCI: notInstalledCI,
    missingDependency: missingDependency,
    invalidOS: invalidOS,
    invalidSmokeTestDisplayError: invalidSmokeTestDisplayError,
    versionMismatch: versionMismatch,
    binaryNotExecutable: binaryNotExecutable,
    unexpected: unexpected,
    failedDownload: failedDownload,
    failedUnzip: failedUnzip,
    failedUnzipWindowsMaxPathLength: failedUnzipWindowsMaxPathLength,
    invalidCypressEnv: invalidCypressEnv,
    invalidCacheDirectory: invalidCacheDirectory,
    CYPRESS_RUN_BINARY: CYPRESS_RUN_BINARY,
    smokeTestFailure: smokeTestFailure,
    childProcessKilled: childProcessKilled,
    incompatibleHeadlessFlags: incompatibleHeadlessFlags,
    invalidRunProjectPath: invalidRunProjectPath,
    invalidTestingType: invalidTestingType,
    incompatibleTestTypeFlags: incompatibleTestTypeFlags,
    incompatibleTestingTypeAndFlag: incompatibleTestingTypeAndFlag,
    invalidConfigFile: invalidConfigFile
  }
};
