// @redis/client/dist/index.js
"use strict";

var __createBinding = void 0 && (void 0).__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);
  if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function get() {
        return m[k];
      }
    };
  }
  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});
var __exportStar = void 0 && (void 0).__exportStar || function (m, exports) {
  for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RedisFlushModes = exports.GeoReplyWith = exports.defineScript = exports.createCluster = exports.commandOptions = exports.createClient = void 0;
var client_1 = require("./lib/client");
var cluster_1 = require("./lib/cluster");
exports.createClient = client_1["default"].create;
exports.commandOptions = client_1["default"].commandOptions;
exports.createCluster = cluster_1["default"].create;
var lua_script_1 = require("./lib/lua-script");
Object.defineProperty(exports, "defineScript", {
  enumerable: true,
  get: function get() {
    return lua_script_1.defineScript;
  }
});
__exportStar(require("./lib/errors"), exports);
var generic_transformers_1 = require("./lib/commands/generic-transformers");
Object.defineProperty(exports, "GeoReplyWith", {
  enumerable: true,
  get: function get() {
    return generic_transformers_1.GeoReplyWith;
  }
});
var FLUSHALL_1 = require("./lib/commands/FLUSHALL");
Object.defineProperty(exports, "RedisFlushModes", {
  enumerable: true,
  get: function get() {
    return FLUSHALL_1.RedisFlushModes;
  }
});
