// @faker-js/faker/dist/esm/chunk-K5VM5AL4.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.b = exports.a = void 0;
var _chunkTN4M2PHH = require("./chunk-TN4M2PHH.mjs");
var _chunk2JBE4HP = require("./chunk-2JBE4HP3.mjs");
var n = ["Boot", "I", "Nu", "Nur", "Tu", "Um", "a", "becoose-a", "boot", "bork", "burn", "chuuses", "cumplete-a", "cun", "cunseqooences", "curcoomstunces", "dee", "deeslikes", "denuoonceeng", "desures", "du", "eccuoont", "ectooel", "edfuntege-a", "efueeds", "egeeen", "ell", "ere-a", "feend", "foolt", "frum", "geefe-a", "gesh", "greet", "heem", "heppeeness", "hes", "hoo", "hoomun", "idea", "ifer", "in", "incuoonter", "injuy", "itselff", "ixcept", "ixemple-a", "ixerceese-a", "ixpleeen", "ixplurer", "ixpuoond", "ixtremely", "knoo", "lebureeuoos", "lufes", "meestekee", "mester-booeelder", "moost", "mun", "nu", "nut", "oobteeen", "oocceseeunelly", "ooccoor", "ooff", "oone-a", "oor", "peeen", "peeenffool", "physeecel", "pleesoore-a", "poorsooe-a", "poorsooes", "preeesing", "prucoore-a", "prudooces", "reeght", "reshunelly", "resooltunt", "sume-a", "teecheengs", "teke-a", "thees", "thet", "thuse-a", "treefiel", "troot", "tu", "tueel", "und", "undertekes", "unnuyeeng", "uny", "unyune-a", "us", "veell", "veet", "ves", "vheech", "vhu", "yuoo", "zee", "zeere-a"];
var l = {
    words: n
  },
  r = l;
var m = {
    title: "English (Bork)",
    code: "en_BORK",
    variant: "BORK",
    language: "en",
    endonym: "English (Bork)",
    dir: "ltr",
    script: "Latn"
  },
  i = m;
var s = [{
  value: "{{person.last_name}}",
  weight: 95
}, {
  value: "{{person.last_name}}-{{person.last_name}}",
  weight: 5
}];
var p = {
    last_name_pattern: s
  },
  a = p;
var f = {
    lorem: r,
    metadata: i,
    person: a
  },
  u = f;
exports.a = u;
var O = new _chunk2JBE4HP.h({
  locale: [u, _chunkTN4M2PHH.b, _chunk2JBE4HP.i]
});
exports.b = O;
