// @faker-js/faker/dist/esm/chunk-Y2I3B4VS.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.b = exports.a = void 0;
var _chunkBQCKYTMH = require("./chunk-BQCKYTMH.mjs");
var _chunkTN4M2PHH = require("./chunk-TN4M2PHH.mjs");
var _chunk2JBE4HP = require("./chunk-2JBE4HP3.mjs");
var i = ["621 ### ###", "661 ### ###", "671 ### ###", "691 ### ###", "+352 621 ### ###", "+352 661 ### ###", "+352 671 ### ###", "+352 691 ### ###"];
var y = {
    formats: i
  },
  n = y;
var m = ["lu"];
var L = {
    domain_suffix: m
  },
  a = L;
var f = ["Diekirch", "Differdange", "Dudelange", "Echternach", "Esch-sur-Alzette", "Ettelbruck", "Grevenmacher", "Luxembourg", "Remich", "Rumelange", "Vianden", "Wiltz"];
var p = ["{{location.city_name}}"];
var l = ["####"];
var u = ["Capellen", "Clervaux", "Diekirch", "Echternach", "Esch-sur-Alzette", "Grevenmacher", "Luxembourg", "Mersch", "Redange", "Remich", "Vianden", "Wiltz"];
var g = {
    city_name: f,
    city_pattern: p,
    postcode: l,
    state: u
  },
  c = g;
var b = {
    title: "French (Luxembourg)",
    code: "fr_LU",
    country: "LU",
    language: "fr",
    endonym: "Fran\xE7ais (Luxembourg)",
    dir: "ltr",
    script: "Latn"
  },
  d = b;
var s = [{
  value: "{{person.last_name}}",
  weight: 1
}];
var P = {
    last_name_pattern: s
  },
  x = P;
var h = ["######", "########", "+352 ######", "+352 ########"];
var k = {
    formats: h
  },
  D = k;
var E = {
    cell_phone: n,
    internet: a,
    location: c,
    metadata: d,
    person: x,
    phone_number: D
  },
  _ = E;
exports.a = _;
var mt = new _chunk2JBE4HP.h({
  locale: [_, _chunkBQCKYTMH.a, _chunkTN4M2PHH.b, _chunk2JBE4HP.i]
});
exports.b = mt;
