// @faker-js/faker/dist/esm/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Aircraft", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.c;
  }
});
Object.defineProperty(exports, "CssFunction", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.e;
  }
});
Object.defineProperty(exports, "CssSpace", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.d;
  }
});
Object.defineProperty(exports, "Faker", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.h;
  }
});
Object.defineProperty(exports, "FakerError", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.b;
  }
});
Object.defineProperty(exports, "Sex", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.f;
  }
});
Object.defineProperty(exports, "af_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkW3RBT6HJ.a;
  }
});
exports.allLocales = exports.allFakers = void 0;
Object.defineProperty(exports, "ar", {
  enumerable: true,
  get: function get() {
    return _chunkYZQFUYRL.a;
  }
});
Object.defineProperty(exports, "az", {
  enumerable: true,
  get: function get() {
    return _chunkW4MH4RRC.a;
  }
});
Object.defineProperty(exports, "base", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.i;
  }
});
Object.defineProperty(exports, "cs_CZ", {
  enumerable: true,
  get: function get() {
    return _chunkJGF6HUZP.a;
  }
});
Object.defineProperty(exports, "de", {
  enumerable: true,
  get: function get() {
    return _chunk7CRZIOOX.a;
  }
});
Object.defineProperty(exports, "de_AT", {
  enumerable: true,
  get: function get() {
    return _chunkPMXLENXA.a;
  }
});
Object.defineProperty(exports, "de_CH", {
  enumerable: true,
  get: function get() {
    return _chunkXY6DNTJF.a;
  }
});
Object.defineProperty(exports, "dv", {
  enumerable: true,
  get: function get() {
    return _chunkZ26WSZSP.a;
  }
});
Object.defineProperty(exports, "el", {
  enumerable: true,
  get: function get() {
    return _chunk4CKQGX5D.a;
  }
});
Object.defineProperty(exports, "en", {
  enumerable: true,
  get: function get() {
    return _chunkTN4M2PHH.b;
  }
});
Object.defineProperty(exports, "en_AU", {
  enumerable: true,
  get: function get() {
    return _chunkTK3SRZ.a;
  }
});
Object.defineProperty(exports, "en_AU_ocker", {
  enumerable: true,
  get: function get() {
    return _chunkABAVZYC.a;
  }
});
Object.defineProperty(exports, "en_BORK", {
  enumerable: true,
  get: function get() {
    return _chunkK5VM5AL.a;
  }
});
Object.defineProperty(exports, "en_CA", {
  enumerable: true,
  get: function get() {
    return _chunkLMN64XYU.a;
  }
});
Object.defineProperty(exports, "en_GB", {
  enumerable: true,
  get: function get() {
    return _chunk4PBAXW2P.a;
  }
});
Object.defineProperty(exports, "en_GH", {
  enumerable: true,
  get: function get() {
    return _chunkDIK7DHCQ.a;
  }
});
Object.defineProperty(exports, "en_HK", {
  enumerable: true,
  get: function get() {
    return _chunkJLGKJLYJ.a;
  }
});
Object.defineProperty(exports, "en_IE", {
  enumerable: true,
  get: function get() {
    return _chunkHXWOMRLC.a;
  }
});
Object.defineProperty(exports, "en_IN", {
  enumerable: true,
  get: function get() {
    return _chunkMGRE6WT.a;
  }
});
Object.defineProperty(exports, "en_NG", {
  enumerable: true,
  get: function get() {
    return _chunkAL22A5IY.a;
  }
});
Object.defineProperty(exports, "en_US", {
  enumerable: true,
  get: function get() {
    return _chunkJVMSY5PD.a;
  }
});
Object.defineProperty(exports, "en_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkGM7LNQLO.a;
  }
});
Object.defineProperty(exports, "es", {
  enumerable: true,
  get: function get() {
    return _chunk6XDWEJQU.a;
  }
});
Object.defineProperty(exports, "es_MX", {
  enumerable: true,
  get: function get() {
    return _chunkL6TUO2GW.a;
  }
});
Object.defineProperty(exports, "fa", {
  enumerable: true,
  get: function get() {
    return _chunkBWQ3DNSA.a;
  }
});
Object.defineProperty(exports, "faker", {
  enumerable: true,
  get: function get() {
    return _chunkMISN5AXR.a;
  }
});
Object.defineProperty(exports, "fakerAF_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkW3RBT6HJ.b;
  }
});
Object.defineProperty(exports, "fakerAR", {
  enumerable: true,
  get: function get() {
    return _chunkYZQFUYRL.b;
  }
});
Object.defineProperty(exports, "fakerAZ", {
  enumerable: true,
  get: function get() {
    return _chunkW4MH4RRC.b;
  }
});
Object.defineProperty(exports, "fakerBASE", {
  enumerable: true,
  get: function get() {
    return _chunkZQUO5W3B.a;
  }
});
Object.defineProperty(exports, "fakerCS_CZ", {
  enumerable: true,
  get: function get() {
    return _chunkJGF6HUZP.b;
  }
});
Object.defineProperty(exports, "fakerDE", {
  enumerable: true,
  get: function get() {
    return _chunk2ALQ655Z.a;
  }
});
Object.defineProperty(exports, "fakerDE_AT", {
  enumerable: true,
  get: function get() {
    return _chunkPMXLENXA.b;
  }
});
Object.defineProperty(exports, "fakerDE_CH", {
  enumerable: true,
  get: function get() {
    return _chunkXY6DNTJF.b;
  }
});
Object.defineProperty(exports, "fakerDV", {
  enumerable: true,
  get: function get() {
    return _chunkZ26WSZSP.b;
  }
});
Object.defineProperty(exports, "fakerEL", {
  enumerable: true,
  get: function get() {
    return _chunk4CKQGX5D.b;
  }
});
Object.defineProperty(exports, "fakerEN", {
  enumerable: true,
  get: function get() {
    return _chunkMISN5AXR.a;
  }
});
Object.defineProperty(exports, "fakerEN_AU", {
  enumerable: true,
  get: function get() {
    return _chunk3NASLQYS.a;
  }
});
Object.defineProperty(exports, "fakerEN_AU_ocker", {
  enumerable: true,
  get: function get() {
    return _chunkABAVZYC.b;
  }
});
Object.defineProperty(exports, "fakerEN_BORK", {
  enumerable: true,
  get: function get() {
    return _chunkK5VM5AL.b;
  }
});
Object.defineProperty(exports, "fakerEN_CA", {
  enumerable: true,
  get: function get() {
    return _chunkLMN64XYU.b;
  }
});
Object.defineProperty(exports, "fakerEN_GB", {
  enumerable: true,
  get: function get() {
    return _chunk4PBAXW2P.b;
  }
});
Object.defineProperty(exports, "fakerEN_GH", {
  enumerable: true,
  get: function get() {
    return _chunkDIK7DHCQ.b;
  }
});
Object.defineProperty(exports, "fakerEN_HK", {
  enumerable: true,
  get: function get() {
    return _chunkJLGKJLYJ.b;
  }
});
Object.defineProperty(exports, "fakerEN_IE", {
  enumerable: true,
  get: function get() {
    return _chunkHXWOMRLC.b;
  }
});
Object.defineProperty(exports, "fakerEN_IN", {
  enumerable: true,
  get: function get() {
    return _chunkMGRE6WT.b;
  }
});
Object.defineProperty(exports, "fakerEN_NG", {
  enumerable: true,
  get: function get() {
    return _chunkAL22A5IY.b;
  }
});
Object.defineProperty(exports, "fakerEN_US", {
  enumerable: true,
  get: function get() {
    return _chunkJVMSY5PD.b;
  }
});
Object.defineProperty(exports, "fakerEN_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkGM7LNQLO.b;
  }
});
Object.defineProperty(exports, "fakerES", {
  enumerable: true,
  get: function get() {
    return _chunkL6A7VLNR.a;
  }
});
Object.defineProperty(exports, "fakerES_MX", {
  enumerable: true,
  get: function get() {
    return _chunkL6TUO2GW.b;
  }
});
Object.defineProperty(exports, "fakerFA", {
  enumerable: true,
  get: function get() {
    return _chunkBWQ3DNSA.b;
  }
});
Object.defineProperty(exports, "fakerFI", {
  enumerable: true,
  get: function get() {
    return _chunk2QFNZSXC.b;
  }
});
Object.defineProperty(exports, "fakerFR", {
  enumerable: true,
  get: function get() {
    return _chunkMV4EHIXT.a;
  }
});
Object.defineProperty(exports, "fakerFR_BE", {
  enumerable: true,
  get: function get() {
    return _chunkQQF4OCYQ.b;
  }
});
Object.defineProperty(exports, "fakerFR_CA", {
  enumerable: true,
  get: function get() {
    return _chunkKD3L7IGL.b;
  }
});
Object.defineProperty(exports, "fakerFR_CH", {
  enumerable: true,
  get: function get() {
    return _chunkCUGUSPRN.b;
  }
});
Object.defineProperty(exports, "fakerFR_LU", {
  enumerable: true,
  get: function get() {
    return _chunkY2I3B4VS.b;
  }
});
Object.defineProperty(exports, "fakerHE", {
  enumerable: true,
  get: function get() {
    return _chunkYXVDPAT.b;
  }
});
Object.defineProperty(exports, "fakerHR", {
  enumerable: true,
  get: function get() {
    return _chunkMVHTNNRQ.b;
  }
});
Object.defineProperty(exports, "fakerHU", {
  enumerable: true,
  get: function get() {
    return _chunkWET6QHOE.b;
  }
});
Object.defineProperty(exports, "fakerHY", {
  enumerable: true,
  get: function get() {
    return _chunkDFPBMR7U.b;
  }
});
Object.defineProperty(exports, "fakerID_ID", {
  enumerable: true,
  get: function get() {
    return _chunkWNQ6EYWU.b;
  }
});
Object.defineProperty(exports, "fakerIT", {
  enumerable: true,
  get: function get() {
    return _chunkGVIEDEAR.b;
  }
});
Object.defineProperty(exports, "fakerJA", {
  enumerable: true,
  get: function get() {
    return _chunkW25WFPAO.b;
  }
});
Object.defineProperty(exports, "fakerKA_GE", {
  enumerable: true,
  get: function get() {
    return _chunkVFMNETMP.b;
  }
});
Object.defineProperty(exports, "fakerKO", {
  enumerable: true,
  get: function get() {
    return _chunkIIORTYPY.b;
  }
});
Object.defineProperty(exports, "fakerLV", {
  enumerable: true,
  get: function get() {
    return _chunk4SRZ7EUR.b;
  }
});
Object.defineProperty(exports, "fakerMK", {
  enumerable: true,
  get: function get() {
    return _chunkTD4WQTZ.b;
  }
});
Object.defineProperty(exports, "fakerNB_NO", {
  enumerable: true,
  get: function get() {
    return _chunkPG2Q73LE.b;
  }
});
Object.defineProperty(exports, "fakerNE", {
  enumerable: true,
  get: function get() {
    return _chunkEXB2HC7I.b;
  }
});
Object.defineProperty(exports, "fakerNL", {
  enumerable: true,
  get: function get() {
    return _chunkDEARTDRM.a;
  }
});
Object.defineProperty(exports, "fakerNL_BE", {
  enumerable: true,
  get: function get() {
    return _chunkG7RIIDN.b;
  }
});
Object.defineProperty(exports, "fakerPL", {
  enumerable: true,
  get: function get() {
    return _chunkYAU5FUDH.b;
  }
});
Object.defineProperty(exports, "fakerPT_BR", {
  enumerable: true,
  get: function get() {
    return _chunkUZGXY7KB.b;
  }
});
Object.defineProperty(exports, "fakerPT_PT", {
  enumerable: true,
  get: function get() {
    return _chunkLNXEZHYO.b;
  }
});
Object.defineProperty(exports, "fakerRO", {
  enumerable: true,
  get: function get() {
    return _chunkQDPL2ECP.a;
  }
});
Object.defineProperty(exports, "fakerRO_MD", {
  enumerable: true,
  get: function get() {
    return _chunkOLPZVZQS.b;
  }
});
Object.defineProperty(exports, "fakerRU", {
  enumerable: true,
  get: function get() {
    return _chunkHIKEHRXB.b;
  }
});
Object.defineProperty(exports, "fakerSK", {
  enumerable: true,
  get: function get() {
    return _chunkBAJCEGMK.b;
  }
});
Object.defineProperty(exports, "fakerSR_RS_latin", {
  enumerable: true,
  get: function get() {
    return _chunk2QQVMEKZ.b;
  }
});
Object.defineProperty(exports, "fakerSV", {
  enumerable: true,
  get: function get() {
    return _chunkQLVJCHTG.b;
  }
});
Object.defineProperty(exports, "fakerTH", {
  enumerable: true,
  get: function get() {
    return _chunkXHCLR24A.b;
  }
});
Object.defineProperty(exports, "fakerTR", {
  enumerable: true,
  get: function get() {
    return _chunkT6RSHQHX.b;
  }
});
Object.defineProperty(exports, "fakerUK", {
  enumerable: true,
  get: function get() {
    return _chunkWEEMGS7B.b;
  }
});
Object.defineProperty(exports, "fakerUR", {
  enumerable: true,
  get: function get() {
    return _chunk2UGI55CL.b;
  }
});
Object.defineProperty(exports, "fakerVI", {
  enumerable: true,
  get: function get() {
    return _chunkAOLOALAF.b;
  }
});
Object.defineProperty(exports, "fakerZH_CN", {
  enumerable: true,
  get: function get() {
    return _chunkZVFD5DJO.b;
  }
});
Object.defineProperty(exports, "fakerZH_TW", {
  enumerable: true,
  get: function get() {
    return _chunkKP77OYU.b;
  }
});
Object.defineProperty(exports, "fakerZU_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkBFO2MQ2Q.b;
  }
});
Object.defineProperty(exports, "fi", {
  enumerable: true,
  get: function get() {
    return _chunk2QFNZSXC.a;
  }
});
Object.defineProperty(exports, "fr", {
  enumerable: true,
  get: function get() {
    return _chunkBQCKYTMH.a;
  }
});
Object.defineProperty(exports, "fr_BE", {
  enumerable: true,
  get: function get() {
    return _chunkQQF4OCYQ.a;
  }
});
Object.defineProperty(exports, "fr_CA", {
  enumerable: true,
  get: function get() {
    return _chunkKD3L7IGL.a;
  }
});
Object.defineProperty(exports, "fr_CH", {
  enumerable: true,
  get: function get() {
    return _chunkCUGUSPRN.a;
  }
});
Object.defineProperty(exports, "fr_LU", {
  enumerable: true,
  get: function get() {
    return _chunkY2I3B4VS.a;
  }
});
Object.defineProperty(exports, "he", {
  enumerable: true,
  get: function get() {
    return _chunkYXVDPAT.a;
  }
});
Object.defineProperty(exports, "hr", {
  enumerable: true,
  get: function get() {
    return _chunkMVHTNNRQ.a;
  }
});
Object.defineProperty(exports, "hu", {
  enumerable: true,
  get: function get() {
    return _chunkWET6QHOE.a;
  }
});
Object.defineProperty(exports, "hy", {
  enumerable: true,
  get: function get() {
    return _chunkDFPBMR7U.a;
  }
});
Object.defineProperty(exports, "id_ID", {
  enumerable: true,
  get: function get() {
    return _chunkWNQ6EYWU.a;
  }
});
Object.defineProperty(exports, "it", {
  enumerable: true,
  get: function get() {
    return _chunkGVIEDEAR.a;
  }
});
Object.defineProperty(exports, "ja", {
  enumerable: true,
  get: function get() {
    return _chunkW25WFPAO.a;
  }
});
Object.defineProperty(exports, "ka_GE", {
  enumerable: true,
  get: function get() {
    return _chunkVFMNETMP.a;
  }
});
Object.defineProperty(exports, "ko", {
  enumerable: true,
  get: function get() {
    return _chunkIIORTYPY.a;
  }
});
Object.defineProperty(exports, "lv", {
  enumerable: true,
  get: function get() {
    return _chunk4SRZ7EUR.a;
  }
});
Object.defineProperty(exports, "mergeLocales", {
  enumerable: true,
  get: function get() {
    return _chunk2JBE4HP.g;
  }
});
Object.defineProperty(exports, "mk", {
  enumerable: true,
  get: function get() {
    return _chunkTD4WQTZ.a;
  }
});
Object.defineProperty(exports, "nb_NO", {
  enumerable: true,
  get: function get() {
    return _chunkPG2Q73LE.a;
  }
});
Object.defineProperty(exports, "ne", {
  enumerable: true,
  get: function get() {
    return _chunkEXB2HC7I.a;
  }
});
Object.defineProperty(exports, "nl", {
  enumerable: true,
  get: function get() {
    return _chunkF7F43K.a;
  }
});
Object.defineProperty(exports, "nl_BE", {
  enumerable: true,
  get: function get() {
    return _chunkG7RIIDN.a;
  }
});
Object.defineProperty(exports, "pl", {
  enumerable: true,
  get: function get() {
    return _chunkYAU5FUDH.a;
  }
});
Object.defineProperty(exports, "pt_BR", {
  enumerable: true,
  get: function get() {
    return _chunkUZGXY7KB.a;
  }
});
Object.defineProperty(exports, "pt_PT", {
  enumerable: true,
  get: function get() {
    return _chunkLNXEZHYO.a;
  }
});
Object.defineProperty(exports, "ro", {
  enumerable: true,
  get: function get() {
    return _chunkQUXXAE7Y.a;
  }
});
Object.defineProperty(exports, "ro_MD", {
  enumerable: true,
  get: function get() {
    return _chunkOLPZVZQS.a;
  }
});
Object.defineProperty(exports, "ru", {
  enumerable: true,
  get: function get() {
    return _chunkHIKEHRXB.a;
  }
});
Object.defineProperty(exports, "sk", {
  enumerable: true,
  get: function get() {
    return _chunkBAJCEGMK.a;
  }
});
Object.defineProperty(exports, "sr_RS_latin", {
  enumerable: true,
  get: function get() {
    return _chunk2QQVMEKZ.a;
  }
});
Object.defineProperty(exports, "sv", {
  enumerable: true,
  get: function get() {
    return _chunkQLVJCHTG.a;
  }
});
Object.defineProperty(exports, "th", {
  enumerable: true,
  get: function get() {
    return _chunkXHCLR24A.a;
  }
});
Object.defineProperty(exports, "tr", {
  enumerable: true,
  get: function get() {
    return _chunkT6RSHQHX.a;
  }
});
Object.defineProperty(exports, "uk", {
  enumerable: true,
  get: function get() {
    return _chunkWEEMGS7B.a;
  }
});
Object.defineProperty(exports, "ur", {
  enumerable: true,
  get: function get() {
    return _chunk2UGI55CL.a;
  }
});
Object.defineProperty(exports, "vi", {
  enumerable: true,
  get: function get() {
    return _chunkAOLOALAF.a;
  }
});
Object.defineProperty(exports, "zh_CN", {
  enumerable: true,
  get: function get() {
    return _chunkZVFD5DJO.a;
  }
});
Object.defineProperty(exports, "zh_TW", {
  enumerable: true,
  get: function get() {
    return _chunkKP77OYU.a;
  }
});
Object.defineProperty(exports, "zu_ZA", {
  enumerable: true,
  get: function get() {
    return _chunkBFO2MQ2Q.a;
  }
});
var _chunkBFO2MQ2Q = require("./chunk-BFO2MQ2Q.mjs");
var _chunkQLVJCHTG = require("./chunk-QLVJCHTG.mjs");
var _chunkXHCLR24A = require("./chunk-XHCLR24A.mjs");
var _chunkT6RSHQHX = require("./chunk-T6RSHQHX.mjs");
var _chunkWEEMGS7B = require("./chunk-WEEMGS7B.mjs");
var _chunk2UGI55CL = require("./chunk-2UGI55CL.mjs");
var _chunkAOLOALAF = require("./chunk-AOLOALAF.mjs");
var _chunkZVFD5DJO = require("./chunk-ZVFD5DJO.mjs");
var _chunkKP77OYU = require("./chunk-KP77OYU7.mjs");
var _chunkYAU5FUDH = require("./chunk-YAU5FUDH.mjs");
var _chunkUZGXY7KB = require("./chunk-UZGXY7KB.mjs");
var _chunkLNXEZHYO = require("./chunk-LNXEZHYO.mjs");
var _chunkQDPL2ECP = require("./chunk-QDPL2ECP.mjs");
var _chunkOLPZVZQS = require("./chunk-OLPZVZQS.mjs");
var _chunkQUXXAE7Y = require("./chunk-QUXXAE7Y.mjs");
var _chunkHIKEHRXB = require("./chunk-HIKEHRXB.mjs");
var _chunkBAJCEGMK = require("./chunk-BAJCEGMK.mjs");
var _chunk2QQVMEKZ = require("./chunk-2QQVMEKZ.mjs");
var _chunkVFMNETMP = require("./chunk-VFMNETMP.mjs");
var _chunkIIORTYPY = require("./chunk-IIORTYPY.mjs");
var _chunk4SRZ7EUR = require("./chunk-4SRZ7EUR.mjs");
var _chunkTD4WQTZ = require("./chunk-TD4WQTZ2.mjs");
var _chunkPG2Q73LE = require("./chunk-PG2Q73LE.mjs");
var _chunkEXB2HC7I = require("./chunk-EXB2HC7I.mjs");
var _chunkDEARTDRM = require("./chunk-DEARTDRM.mjs");
var _chunkG7RIIDN = require("./chunk-G7RIIDN3.mjs");
var _chunkF7F43K = require("./chunk-F7F43K56.mjs");
var _chunkY2I3B4VS = require("./chunk-Y2I3B4VS.mjs");
var _chunkYXVDPAT = require("./chunk-YXVDPAT5.mjs");
var _chunkMVHTNNRQ = require("./chunk-MVHTNNRQ.mjs");
var _chunkWET6QHOE = require("./chunk-WET6QHOE.mjs");
var _chunkDFPBMR7U = require("./chunk-DFPBMR7U.mjs");
var _chunkWNQ6EYWU = require("./chunk-WNQ6EYWU.mjs");
var _chunkGVIEDEAR = require("./chunk-GVIEDEAR.mjs");
var _chunkW25WFPAO = require("./chunk-W25WFPAO.mjs");
var _chunkL6A7VLNR = require("./chunk-L6A7VLNR.mjs");
var _chunkL6TUO2GW = require("./chunk-L6TUO2GW.mjs");
var _chunk6XDWEJQU = require("./chunk-6XDWEJQU.mjs");
var _chunkBWQ3DNSA = require("./chunk-BWQ3DNSA.mjs");
var _chunk2QFNZSXC = require("./chunk-2QFNZSXC.mjs");
var _chunkMV4EHIXT = require("./chunk-MV4EHIXT.mjs");
var _chunkQQF4OCYQ = require("./chunk-QQF4OCYQ.mjs");
var _chunkKD3L7IGL = require("./chunk-KD3L7IGL.mjs");
var _chunkCUGUSPRN = require("./chunk-CUGUSPRN.mjs");
var _chunkBQCKYTMH = require("./chunk-BQCKYTMH.mjs");
var _chunk4PBAXW2P = require("./chunk-4PBAXW2P.mjs");
var _chunkDIK7DHCQ = require("./chunk-DIK7DHCQ.mjs");
var _chunkJLGKJLYJ = require("./chunk-JLGKJLYJ.mjs");
var _chunkHXWOMRLC = require("./chunk-HXWOMRLC.mjs");
var _chunkMGRE6WT = require("./chunk-MGRE6WT7.mjs");
var _chunkAL22A5IY = require("./chunk-AL22A5IY.mjs");
var _chunkJVMSY5PD = require("./chunk-JVMSY5PD.mjs");
var _chunkGM7LNQLO = require("./chunk-GM7LNQLO.mjs");
var _chunkXY6DNTJF = require("./chunk-XY6DNTJF.mjs");
var _chunkZ26WSZSP = require("./chunk-Z26WSZSP.mjs");
var _chunk4CKQGX5D = require("./chunk-4CKQGX5D.mjs");
var _chunkMISN5AXR = require("./chunk-MISN5AXR.mjs");
var _chunk3NASLQYS = require("./chunk-3NASLQYS.mjs");
var _chunkABAVZYC = require("./chunk-ABAVZYC4.mjs");
var _chunkTK3SRZ = require("./chunk-TK3SRZ37.mjs");
var _chunkK5VM5AL = require("./chunk-K5VM5AL4.mjs");
var _chunkLMN64XYU = require("./chunk-LMN64XYU.mjs");
var _chunkW3RBT6HJ = require("./chunk-W3RBT6HJ.mjs");
var _chunkYZQFUYRL = require("./chunk-YZQFUYRL.mjs");
var _chunkW4MH4RRC = require("./chunk-W4MH4RRC.mjs");
var _chunkZQUO5W3B = require("./chunk-ZQUO5W3B.mjs");
var _chunkJGF6HUZP = require("./chunk-JGF6HUZP.mjs");
var _chunk2ALQ655Z = require("./chunk-2ALQ655Z.mjs");
var _chunkPMXLENXA = require("./chunk-PMXLENXA.mjs");
var _chunk7CRZIOOX = require("./chunk-7CRZIOOX.mjs");
var _chunkTN4M2PHH = require("./chunk-TN4M2PHH.mjs");
var _chunk2JBE4HP = require("./chunk-2JBE4HP3.mjs");
var Oo = {
  af_ZA: _chunkW3RBT6HJ.b,
  ar: _chunkYZQFUYRL.b,
  az: _chunkW4MH4RRC.b,
  base: _chunkZQUO5W3B.a,
  cs_CZ: _chunkJGF6HUZP.b,
  de: _chunk2ALQ655Z.a,
  de_AT: _chunkPMXLENXA.b,
  de_CH: _chunkXY6DNTJF.b,
  dv: _chunkZ26WSZSP.b,
  el: _chunk4CKQGX5D.b,
  en: _chunkMISN5AXR.a,
  en_AU: _chunk3NASLQYS.a,
  en_AU_ocker: _chunkABAVZYC.b,
  en_BORK: _chunkK5VM5AL.b,
  en_CA: _chunkLMN64XYU.b,
  en_GB: _chunk4PBAXW2P.b,
  en_GH: _chunkDIK7DHCQ.b,
  en_HK: _chunkJLGKJLYJ.b,
  en_IE: _chunkHXWOMRLC.b,
  en_IN: _chunkMGRE6WT.b,
  en_NG: _chunkAL22A5IY.b,
  en_US: _chunkJVMSY5PD.b,
  en_ZA: _chunkGM7LNQLO.b,
  es: _chunkL6A7VLNR.a,
  es_MX: _chunkL6TUO2GW.b,
  fa: _chunkBWQ3DNSA.b,
  fi: _chunk2QFNZSXC.b,
  fr: _chunkMV4EHIXT.a,
  fr_BE: _chunkQQF4OCYQ.b,
  fr_CA: _chunkKD3L7IGL.b,
  fr_CH: _chunkCUGUSPRN.b,
  fr_LU: _chunkY2I3B4VS.b,
  he: _chunkYXVDPAT.b,
  hr: _chunkMVHTNNRQ.b,
  hu: _chunkWET6QHOE.b,
  hy: _chunkDFPBMR7U.b,
  id_ID: _chunkWNQ6EYWU.b,
  it: _chunkGVIEDEAR.b,
  ja: _chunkW25WFPAO.b,
  ka_GE: _chunkVFMNETMP.b,
  ko: _chunkIIORTYPY.b,
  lv: _chunk4SRZ7EUR.b,
  mk: _chunkTD4WQTZ.b,
  nb_NO: _chunkPG2Q73LE.b,
  ne: _chunkEXB2HC7I.b,
  nl: _chunkDEARTDRM.a,
  nl_BE: _chunkG7RIIDN.b,
  pl: _chunkYAU5FUDH.b,
  pt_BR: _chunkUZGXY7KB.b,
  pt_PT: _chunkLNXEZHYO.b,
  ro: _chunkQDPL2ECP.a,
  ro_MD: _chunkOLPZVZQS.b,
  ru: _chunkHIKEHRXB.b,
  sk: _chunkBAJCEGMK.b,
  sr_RS_latin: _chunk2QQVMEKZ.b,
  sv: _chunkQLVJCHTG.b,
  th: _chunkXHCLR24A.b,
  tr: _chunkT6RSHQHX.b,
  uk: _chunkWEEMGS7B.b,
  ur: _chunk2UGI55CL.b,
  vi: _chunkAOLOALAF.b,
  zh_CN: _chunkZVFD5DJO.b,
  zh_TW: _chunkKP77OYU.b,
  zu_ZA: _chunkBFO2MQ2Q.b
};
exports.allFakers = Oo;
var r = {};
exports.allLocales = r;
(0, _chunk2JBE4HP.a)(r, {
  af_ZA: function af_ZA() {
    return _chunkW3RBT6HJ.a;
  },
  ar: function ar() {
    return _chunkYZQFUYRL.a;
  },
  az: function az() {
    return _chunkW4MH4RRC.a;
  },
  base: function base() {
    return _chunk2JBE4HP.i;
  },
  cs_CZ: function cs_CZ() {
    return _chunkJGF6HUZP.a;
  },
  de: function de() {
    return _chunk7CRZIOOX.a;
  },
  de_AT: function de_AT() {
    return _chunkPMXLENXA.a;
  },
  de_CH: function de_CH() {
    return _chunkXY6DNTJF.a;
  },
  dv: function dv() {
    return _chunkZ26WSZSP.a;
  },
  el: function el() {
    return _chunk4CKQGX5D.a;
  },
  en: function en() {
    return _chunkTN4M2PHH.b;
  },
  en_AU: function en_AU() {
    return _chunkTK3SRZ.a;
  },
  en_AU_ocker: function en_AU_ocker() {
    return _chunkABAVZYC.a;
  },
  en_BORK: function en_BORK() {
    return _chunkK5VM5AL.a;
  },
  en_CA: function en_CA() {
    return _chunkLMN64XYU.a;
  },
  en_GB: function en_GB() {
    return _chunk4PBAXW2P.a;
  },
  en_GH: function en_GH() {
    return _chunkDIK7DHCQ.a;
  },
  en_HK: function en_HK() {
    return _chunkJLGKJLYJ.a;
  },
  en_IE: function en_IE() {
    return _chunkHXWOMRLC.a;
  },
  en_IN: function en_IN() {
    return _chunkMGRE6WT.a;
  },
  en_NG: function en_NG() {
    return _chunkAL22A5IY.a;
  },
  en_US: function en_US() {
    return _chunkJVMSY5PD.a;
  },
  en_ZA: function en_ZA() {
    return _chunkGM7LNQLO.a;
  },
  es: function es() {
    return _chunk6XDWEJQU.a;
  },
  es_MX: function es_MX() {
    return _chunkL6TUO2GW.a;
  },
  fa: function fa() {
    return _chunkBWQ3DNSA.a;
  },
  fi: function fi() {
    return _chunk2QFNZSXC.a;
  },
  fr: function fr() {
    return _chunkBQCKYTMH.a;
  },
  fr_BE: function fr_BE() {
    return _chunkQQF4OCYQ.a;
  },
  fr_CA: function fr_CA() {
    return _chunkKD3L7IGL.a;
  },
  fr_CH: function fr_CH() {
    return _chunkCUGUSPRN.a;
  },
  fr_LU: function fr_LU() {
    return _chunkY2I3B4VS.a;
  },
  he: function he() {
    return _chunkYXVDPAT.a;
  },
  hr: function hr() {
    return _chunkMVHTNNRQ.a;
  },
  hu: function hu() {
    return _chunkWET6QHOE.a;
  },
  hy: function hy() {
    return _chunkDFPBMR7U.a;
  },
  id_ID: function id_ID() {
    return _chunkWNQ6EYWU.a;
  },
  it: function it() {
    return _chunkGVIEDEAR.a;
  },
  ja: function ja() {
    return _chunkW25WFPAO.a;
  },
  ka_GE: function ka_GE() {
    return _chunkVFMNETMP.a;
  },
  ko: function ko() {
    return _chunkIIORTYPY.a;
  },
  lv: function lv() {
    return _chunk4SRZ7EUR.a;
  },
  mk: function mk() {
    return _chunkTD4WQTZ.a;
  },
  nb_NO: function nb_NO() {
    return _chunkPG2Q73LE.a;
  },
  ne: function ne() {
    return _chunkEXB2HC7I.a;
  },
  nl: function nl() {
    return _chunkF7F43K.a;
  },
  nl_BE: function nl_BE() {
    return _chunkG7RIIDN.a;
  },
  pl: function pl() {
    return _chunkYAU5FUDH.a;
  },
  pt_BR: function pt_BR() {
    return _chunkUZGXY7KB.a;
  },
  pt_PT: function pt_PT() {
    return _chunkLNXEZHYO.a;
  },
  ro: function ro() {
    return _chunkQUXXAE7Y.a;
  },
  ro_MD: function ro_MD() {
    return _chunkOLPZVZQS.a;
  },
  ru: function ru() {
    return _chunkHIKEHRXB.a;
  },
  sk: function sk() {
    return _chunkBAJCEGMK.a;
  },
  sr_RS_latin: function sr_RS_latin() {
    return _chunk2QQVMEKZ.a;
  },
  sv: function sv() {
    return _chunkQLVJCHTG.a;
  },
  th: function th() {
    return _chunkXHCLR24A.a;
  },
  tr: function tr() {
    return _chunkT6RSHQHX.a;
  },
  uk: function uk() {
    return _chunkWEEMGS7B.a;
  },
  ur: function ur() {
    return _chunk2UGI55CL.a;
  },
  vi: function vi() {
    return _chunkAOLOALAF.a;
  },
  zh_CN: function zh_CN() {
    return _chunkZVFD5DJO.a;
  },
  zh_TW: function zh_TW() {
    return _chunkKP77OYU.a;
  },
  zu_ZA: function zu_ZA() {
    return _chunkBFO2MQ2Q.a;
  }
});
