// @redis/client/dist/lib/cluster/commands.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var APPEND = require("../commands/APPEND");
var BITCOUNT = require("../commands/BITCOUNT");
var BITFIELD_RO = require("../commands/BITFIELD_RO");
var BITFIELD = require("../commands/BITFIELD");
var BITOP = require("../commands/BITOP");
var BITPOS = require("../commands/BITPOS");
var BLMOVE = require("../commands/BLMOVE");
var BLMPOP = require("../commands/BLMPOP");
var BLPOP = require("../commands/BLPOP");
var BRPOP = require("../commands/BRPOP");
var BRPOPLPUSH = require("../commands/BRPOPLPUSH");
var BZMPOP = require("../commands/BZMPOP");
var BZPOPMAX = require("../commands/BZPOPMAX");
var BZPOPMIN = require("../commands/BZPOPMIN");
var COPY = require("../commands/COPY");
var DECR = require("../commands/DECR");
var DECRBY = require("../commands/DECRBY");
var DEL = require("../commands/DEL");
var DUMP = require("../commands/DUMP");
var EVAL_RO = require("../commands/EVAL_RO");
var EVAL = require("../commands/EVAL");
var EVALSHA_RO = require("../commands/EVALSHA_RO");
var EVALSHA = require("../commands/EVALSHA");
var EXISTS = require("../commands/EXISTS");
var EXPIRE = require("../commands/EXPIRE");
var EXPIREAT = require("../commands/EXPIREAT");
var EXPIRETIME = require("../commands/EXPIRETIME");
var FCALL_RO = require("../commands/FCALL_RO");
var FCALL = require("../commands/FCALL");
var GEOADD = require("../commands/GEOADD");
var GEODIST = require("../commands/GEODIST");
var GEOHASH = require("../commands/GEOHASH");
var GEOPOS = require("../commands/GEOPOS");
var GEORADIUS_RO_WITH = require("../commands/GEORADIUS_RO_WITH");
var GEORADIUS_RO = require("../commands/GEORADIUS_RO");
var GEORADIUS_WITH = require("../commands/GEORADIUS_WITH");
var GEORADIUS = require("../commands/GEORADIUS");
var GEORADIUSBYMEMBER_RO_WITH = require("../commands/GEORADIUSBYMEMBER_RO_WITH");
var GEORADIUSBYMEMBER_RO = require("../commands/GEORADIUSBYMEMBER_RO");
var GEORADIUSBYMEMBER_WITH = require("../commands/GEORADIUSBYMEMBER_WITH");
var GEORADIUSBYMEMBER = require("../commands/GEORADIUSBYMEMBER");
var GEORADIUSBYMEMBERSTORE = require("../commands/GEORADIUSBYMEMBERSTORE");
var GEORADIUSSTORE = require("../commands/GEORADIUSSTORE");
var GEOSEARCH_WITH = require("../commands/GEOSEARCH_WITH");
var GEOSEARCH = require("../commands/GEOSEARCH");
var GEOSEARCHSTORE = require("../commands/GEOSEARCHSTORE");
var GET = require("../commands/GET");
var GETBIT = require("../commands/GETBIT");
var GETDEL = require("../commands/GETDEL");
var GETEX = require("../commands/GETEX");
var GETRANGE = require("../commands/GETRANGE");
var GETSET = require("../commands/GETSET");
var HDEL = require("../commands/HDEL");
var HEXISTS = require("../commands/HEXISTS");
var HGET = require("../commands/HGET");
var HGETALL = require("../commands/HGETALL");
var HINCRBY = require("../commands/HINCRBY");
var HINCRBYFLOAT = require("../commands/HINCRBYFLOAT");
var HKEYS = require("../commands/HKEYS");
var HLEN = require("../commands/HLEN");
var HMGET = require("../commands/HMGET");
var HRANDFIELD_COUNT_WITHVALUES = require("../commands/HRANDFIELD_COUNT_WITHVALUES");
var HRANDFIELD_COUNT = require("../commands/HRANDFIELD_COUNT");
var HRANDFIELD = require("../commands/HRANDFIELD");
var HSCAN = require("../commands/HSCAN");
var HSET = require("../commands/HSET");
var HSETNX = require("../commands/HSETNX");
var HSTRLEN = require("../commands/HSTRLEN");
var HVALS = require("../commands/HVALS");
var INCR = require("../commands/INCR");
var INCRBY = require("../commands/INCRBY");
var INCRBYFLOAT = require("../commands/INCRBYFLOAT");
var LCS_IDX_WITHMATCHLEN = require("../commands/LCS_IDX_WITHMATCHLEN");
var LCS_IDX = require("../commands/LCS_IDX");
var LCS_LEN = require("../commands/LCS_LEN");
var LCS = require("../commands/LCS");
var LINDEX = require("../commands/LINDEX");
var LINSERT = require("../commands/LINSERT");
var LLEN = require("../commands/LLEN");
var LMOVE = require("../commands/LMOVE");
var LMPOP = require("../commands/LMPOP");
var LPOP_COUNT = require("../commands/LPOP_COUNT");
var LPOP = require("../commands/LPOP");
var LPOS_COUNT = require("../commands/LPOS_COUNT");
var LPOS = require("../commands/LPOS");
var LPUSH = require("../commands/LPUSH");
var LPUSHX = require("../commands/LPUSHX");
var LRANGE = require("../commands/LRANGE");
var LREM = require("../commands/LREM");
var LSET = require("../commands/LSET");
var LTRIM = require("../commands/LTRIM");
var MGET = require("../commands/MGET");
var MIGRATE = require("../commands/MIGRATE");
var MSET = require("../commands/MSET");
var MSETNX = require("../commands/MSETNX");
var OBJECT_ENCODING = require("../commands/OBJECT_ENCODING");
var OBJECT_FREQ = require("../commands/OBJECT_FREQ");
var OBJECT_IDLETIME = require("../commands/OBJECT_IDLETIME");
var OBJECT_REFCOUNT = require("../commands/OBJECT_REFCOUNT");
var PERSIST = require("../commands/PERSIST");
var PEXPIRE = require("../commands/PEXPIRE");
var PEXPIREAT = require("../commands/PEXPIREAT");
var PEXPIRETIME = require("../commands/PEXPIRETIME");
var PFADD = require("../commands/PFADD");
var PFCOUNT = require("../commands/PFCOUNT");
var PFMERGE = require("../commands/PFMERGE");
var PSETEX = require("../commands/PSETEX");
var PTTL = require("../commands/PTTL");
var PUBLISH = require("../commands/PUBLISH");
var RENAME = require("../commands/RENAME");
var RENAMENX = require("../commands/RENAMENX");
var RPOP_COUNT = require("../commands/RPOP_COUNT");
var RPOP = require("../commands/RPOP");
var RPOPLPUSH = require("../commands/RPOPLPUSH");
var RPUSH = require("../commands/RPUSH");
var RPUSHX = require("../commands/RPUSHX");
var SADD = require("../commands/SADD");
var SCARD = require("../commands/SCARD");
var SDIFF = require("../commands/SDIFF");
var SDIFFSTORE = require("../commands/SDIFFSTORE");
var SET = require("../commands/SET");
var SETBIT = require("../commands/SETBIT");
var SETEX = require("../commands/SETEX");
var SETNX = require("../commands/SETNX");
var SETRANGE = require("../commands/SETRANGE");
var SINTER = require("../commands/SINTER");
var SINTERCARD = require("../commands/SINTERCARD");
var SINTERSTORE = require("../commands/SINTERSTORE");
var SISMEMBER = require("../commands/SISMEMBER");
var SMEMBERS = require("../commands/SMEMBERS");
var SMISMEMBER = require("../commands/SMISMEMBER");
var SMOVE = require("../commands/SMOVE");
var SORT_RO = require("../commands/SORT_RO");
var SORT_STORE = require("../commands/SORT_STORE");
var SORT = require("../commands/SORT");
var SPOP = require("../commands/SPOP");
var SPUBLISH = require("../commands/SPUBLISH");
var SRANDMEMBER_COUNT = require("../commands/SRANDMEMBER_COUNT");
var SRANDMEMBER = require("../commands/SRANDMEMBER");
var SREM = require("../commands/SREM");
var SSCAN = require("../commands/SSCAN");
var STRLEN = require("../commands/STRLEN");
var SUNION = require("../commands/SUNION");
var SUNIONSTORE = require("../commands/SUNIONSTORE");
var TOUCH = require("../commands/TOUCH");
var TTL = require("../commands/TTL");
var TYPE = require("../commands/TYPE");
var UNLINK = require("../commands/UNLINK");
var WATCH = require("../commands/WATCH");
var XACK = require("../commands/XACK");
var XADD = require("../commands/XADD");
var XAUTOCLAIM_JUSTID = require("../commands/XAUTOCLAIM_JUSTID");
var XAUTOCLAIM = require("../commands/XAUTOCLAIM");
var XCLAIM_JUSTID = require("../commands/XCLAIM_JUSTID");
var XCLAIM = require("../commands/XCLAIM");
var XDEL = require("../commands/XDEL");
var XGROUP_CREATE = require("../commands/XGROUP_CREATE");
var XGROUP_CREATECONSUMER = require("../commands/XGROUP_CREATECONSUMER");
var XGROUP_DELCONSUMER = require("../commands/XGROUP_DELCONSUMER");
var XGROUP_DESTROY = require("../commands/XGROUP_DESTROY");
var XGROUP_SETID = require("../commands/XGROUP_SETID");
var XINFO_CONSUMERS = require("../commands/XINFO_CONSUMERS");
var XINFO_GROUPS = require("../commands/XINFO_GROUPS");
var XINFO_STREAM = require("../commands/XINFO_STREAM");
var XLEN = require("../commands/XLEN");
var XPENDING_RANGE = require("../commands/XPENDING_RANGE");
var XPENDING = require("../commands/XPENDING");
var XRANGE = require("../commands/XRANGE");
var XREAD = require("../commands/XREAD");
var XREADGROUP = require("../commands/XREADGROUP");
var XREVRANGE = require("../commands/XREVRANGE");
var XSETID = require("../commands/XSETID");
var XTRIM = require("../commands/XTRIM");
var ZADD = require("../commands/ZADD");
var ZCARD = require("../commands/ZCARD");
var ZCOUNT = require("../commands/ZCOUNT");
var ZDIFF_WITHSCORES = require("../commands/ZDIFF_WITHSCORES");
var ZDIFF = require("../commands/ZDIFF");
var ZDIFFSTORE = require("../commands/ZDIFFSTORE");
var ZINCRBY = require("../commands/ZINCRBY");
var ZINTER_WITHSCORES = require("../commands/ZINTER_WITHSCORES");
var ZINTER = require("../commands/ZINTER");
var ZINTERCARD = require("../commands/ZINTERCARD");
var ZINTERSTORE = require("../commands/ZINTERSTORE");
var ZLEXCOUNT = require("../commands/ZLEXCOUNT");
var ZMPOP = require("../commands/ZMPOP");
var ZMSCORE = require("../commands/ZMSCORE");
var ZPOPMAX_COUNT = require("../commands/ZPOPMAX_COUNT");
var ZPOPMAX = require("../commands/ZPOPMAX");
var ZPOPMIN_COUNT = require("../commands/ZPOPMIN_COUNT");
var ZPOPMIN = require("../commands/ZPOPMIN");
var ZRANDMEMBER_COUNT_WITHSCORES = require("../commands/ZRANDMEMBER_COUNT_WITHSCORES");
var ZRANDMEMBER_COUNT = require("../commands/ZRANDMEMBER_COUNT");
var ZRANDMEMBER = require("../commands/ZRANDMEMBER");
var ZRANGE_WITHSCORES = require("../commands/ZRANGE_WITHSCORES");
var ZRANGE = require("../commands/ZRANGE");
var ZRANGEBYLEX = require("../commands/ZRANGEBYLEX");
var ZRANGEBYSCORE_WITHSCORES = require("../commands/ZRANGEBYSCORE_WITHSCORES");
var ZRANGEBYSCORE = require("../commands/ZRANGEBYSCORE");
var ZRANGESTORE = require("../commands/ZRANGESTORE");
var ZRANK = require("../commands/ZRANK");
var ZREM = require("../commands/ZREM");
var ZREMRANGEBYLEX = require("../commands/ZREMRANGEBYLEX");
var ZREMRANGEBYRANK = require("../commands/ZREMRANGEBYRANK");
var ZREMRANGEBYSCORE = require("../commands/ZREMRANGEBYSCORE");
var ZREVRANK = require("../commands/ZREVRANK");
var ZSCAN = require("../commands/ZSCAN");
var ZSCORE = require("../commands/ZSCORE");
var ZUNION_WITHSCORES = require("../commands/ZUNION_WITHSCORES");
var ZUNION = require("../commands/ZUNION");
var ZUNIONSTORE = require("../commands/ZUNIONSTORE");
exports["default"] = {
  APPEND: APPEND,
  append: APPEND,
  BITCOUNT: BITCOUNT,
  bitCount: BITCOUNT,
  BITFIELD_RO: BITFIELD_RO,
  bitFieldRo: BITFIELD_RO,
  BITFIELD: BITFIELD,
  bitField: BITFIELD,
  BITOP: BITOP,
  bitOp: BITOP,
  BITPOS: BITPOS,
  bitPos: BITPOS,
  BLMOVE: BLMOVE,
  blMove: BLMOVE,
  BLMPOP: BLMPOP,
  blmPop: BLMPOP,
  BLPOP: BLPOP,
  blPop: BLPOP,
  BRPOP: BRPOP,
  brPop: BRPOP,
  BRPOPLPUSH: BRPOPLPUSH,
  brPopLPush: BRPOPLPUSH,
  BZMPOP: BZMPOP,
  bzmPop: BZMPOP,
  BZPOPMAX: BZPOPMAX,
  bzPopMax: BZPOPMAX,
  BZPOPMIN: BZPOPMIN,
  bzPopMin: BZPOPMIN,
  COPY: COPY,
  copy: COPY,
  DECR: DECR,
  decr: DECR,
  DECRBY: DECRBY,
  decrBy: DECRBY,
  DEL: DEL,
  del: DEL,
  DUMP: DUMP,
  dump: DUMP,
  EVAL_RO: EVAL_RO,
  evalRo: EVAL_RO,
  EVAL: EVAL,
  eval: EVAL,
  EVALSHA: EVALSHA,
  evalSha: EVALSHA,
  EVALSHA_RO: EVALSHA_RO,
  evalShaRo: EVALSHA_RO,
  EXISTS: EXISTS,
  exists: EXISTS,
  EXPIRE: EXPIRE,
  expire: EXPIRE,
  EXPIREAT: EXPIREAT,
  expireAt: EXPIREAT,
  EXPIRETIME: EXPIRETIME,
  expireTime: EXPIRETIME,
  FCALL_RO: FCALL_RO,
  fCallRo: FCALL_RO,
  FCALL: FCALL,
  fCall: FCALL,
  GEOADD: GEOADD,
  geoAdd: GEOADD,
  GEODIST: GEODIST,
  geoDist: GEODIST,
  GEOHASH: GEOHASH,
  geoHash: GEOHASH,
  GEOPOS: GEOPOS,
  geoPos: GEOPOS,
  GEORADIUS_RO_WITH: GEORADIUS_RO_WITH,
  geoRadiusRoWith: GEORADIUS_RO_WITH,
  GEORADIUS_RO: GEORADIUS_RO,
  geoRadiusRo: GEORADIUS_RO,
  GEORADIUS_WITH: GEORADIUS_WITH,
  geoRadiusWith: GEORADIUS_WITH,
  GEORADIUS: GEORADIUS,
  geoRadius: GEORADIUS,
  GEORADIUSBYMEMBER_RO_WITH: GEORADIUSBYMEMBER_RO_WITH,
  geoRadiusByMemberRoWith: GEORADIUSBYMEMBER_RO_WITH,
  GEORADIUSBYMEMBER_RO: GEORADIUSBYMEMBER_RO,
  geoRadiusByMemberRo: GEORADIUSBYMEMBER_RO,
  GEORADIUSBYMEMBER_WITH: GEORADIUSBYMEMBER_WITH,
  geoRadiusByMemberWith: GEORADIUSBYMEMBER_WITH,
  GEORADIUSBYMEMBER: GEORADIUSBYMEMBER,
  geoRadiusByMember: GEORADIUSBYMEMBER,
  GEORADIUSBYMEMBERSTORE: GEORADIUSBYMEMBERSTORE,
  geoRadiusByMemberStore: GEORADIUSBYMEMBERSTORE,
  GEORADIUSSTORE: GEORADIUSSTORE,
  geoRadiusStore: GEORADIUSSTORE,
  GEOSEARCH_WITH: GEOSEARCH_WITH,
  geoSearchWith: GEOSEARCH_WITH,
  GEOSEARCH: GEOSEARCH,
  geoSearch: GEOSEARCH,
  GEOSEARCHSTORE: GEOSEARCHSTORE,
  geoSearchStore: GEOSEARCHSTORE,
  GET: GET,
  get: GET,
  GETBIT: GETBIT,
  getBit: GETBIT,
  GETDEL: GETDEL,
  getDel: GETDEL,
  GETEX: GETEX,
  getEx: GETEX,
  GETRANGE: GETRANGE,
  getRange: GETRANGE,
  GETSET: GETSET,
  getSet: GETSET,
  HDEL: HDEL,
  hDel: HDEL,
  HEXISTS: HEXISTS,
  hExists: HEXISTS,
  HGET: HGET,
  hGet: HGET,
  HGETALL: HGETALL,
  hGetAll: HGETALL,
  HINCRBY: HINCRBY,
  hIncrBy: HINCRBY,
  HINCRBYFLOAT: HINCRBYFLOAT,
  hIncrByFloat: HINCRBYFLOAT,
  HKEYS: HKEYS,
  hKeys: HKEYS,
  HLEN: HLEN,
  hLen: HLEN,
  HMGET: HMGET,
  hmGet: HMGET,
  HRANDFIELD_COUNT_WITHVALUES: HRANDFIELD_COUNT_WITHVALUES,
  hRandFieldCountWithValues: HRANDFIELD_COUNT_WITHVALUES,
  HRANDFIELD_COUNT: HRANDFIELD_COUNT,
  hRandFieldCount: HRANDFIELD_COUNT,
  HRANDFIELD: HRANDFIELD,
  hRandField: HRANDFIELD,
  HSCAN: HSCAN,
  hScan: HSCAN,
  HSET: HSET,
  hSet: HSET,
  HSETNX: HSETNX,
  hSetNX: HSETNX,
  HSTRLEN: HSTRLEN,
  hStrLen: HSTRLEN,
  HVALS: HVALS,
  hVals: HVALS,
  INCR: INCR,
  incr: INCR,
  INCRBY: INCRBY,
  incrBy: INCRBY,
  INCRBYFLOAT: INCRBYFLOAT,
  incrByFloat: INCRBYFLOAT,
  LCS_IDX_WITHMATCHLEN: LCS_IDX_WITHMATCHLEN,
  lcsIdxWithMatchLen: LCS_IDX_WITHMATCHLEN,
  LCS_IDX: LCS_IDX,
  lcsIdx: LCS_IDX,
  LCS_LEN: LCS_LEN,
  lcsLen: LCS_LEN,
  LCS: LCS,
  lcs: LCS,
  LINDEX: LINDEX,
  lIndex: LINDEX,
  LINSERT: LINSERT,
  lInsert: LINSERT,
  LLEN: LLEN,
  lLen: LLEN,
  LMOVE: LMOVE,
  lMove: LMOVE,
  LMPOP: LMPOP,
  lmPop: LMPOP,
  LPOP_COUNT: LPOP_COUNT,
  lPopCount: LPOP_COUNT,
  LPOP: LPOP,
  lPop: LPOP,
  LPOS_COUNT: LPOS_COUNT,
  lPosCount: LPOS_COUNT,
  LPOS: LPOS,
  lPos: LPOS,
  LPUSH: LPUSH,
  lPush: LPUSH,
  LPUSHX: LPUSHX,
  lPushX: LPUSHX,
  LRANGE: LRANGE,
  lRange: LRANGE,
  LREM: LREM,
  lRem: LREM,
  LSET: LSET,
  lSet: LSET,
  LTRIM: LTRIM,
  lTrim: LTRIM,
  MGET: MGET,
  mGet: MGET,
  MIGRATE: MIGRATE,
  migrate: MIGRATE,
  MSET: MSET,
  mSet: MSET,
  MSETNX: MSETNX,
  mSetNX: MSETNX,
  OBJECT_ENCODING: OBJECT_ENCODING,
  objectEncoding: OBJECT_ENCODING,
  OBJECT_FREQ: OBJECT_FREQ,
  objectFreq: OBJECT_FREQ,
  OBJECT_IDLETIME: OBJECT_IDLETIME,
  objectIdleTime: OBJECT_IDLETIME,
  OBJECT_REFCOUNT: OBJECT_REFCOUNT,
  objectRefCount: OBJECT_REFCOUNT,
  PERSIST: PERSIST,
  persist: PERSIST,
  PEXPIRE: PEXPIRE,
  pExpire: PEXPIRE,
  PEXPIREAT: PEXPIREAT,
  pExpireAt: PEXPIREAT,
  PEXPIRETIME: PEXPIRETIME,
  pExpireTime: PEXPIRETIME,
  PFADD: PFADD,
  pfAdd: PFADD,
  PFCOUNT: PFCOUNT,
  pfCount: PFCOUNT,
  PFMERGE: PFMERGE,
  pfMerge: PFMERGE,
  PSETEX: PSETEX,
  pSetEx: PSETEX,
  PTTL: PTTL,
  pTTL: PTTL,
  PUBLISH: PUBLISH,
  publish: PUBLISH,
  RENAME: RENAME,
  rename: RENAME,
  RENAMENX: RENAMENX,
  renameNX: RENAMENX,
  RPOP_COUNT: RPOP_COUNT,
  rPopCount: RPOP_COUNT,
  RPOP: RPOP,
  rPop: RPOP,
  RPOPLPUSH: RPOPLPUSH,
  rPopLPush: RPOPLPUSH,
  RPUSH: RPUSH,
  rPush: RPUSH,
  RPUSHX: RPUSHX,
  rPushX: RPUSHX,
  SADD: SADD,
  sAdd: SADD,
  SCARD: SCARD,
  sCard: SCARD,
  SDIFF: SDIFF,
  sDiff: SDIFF,
  SDIFFSTORE: SDIFFSTORE,
  sDiffStore: SDIFFSTORE,
  SINTER: SINTER,
  sInter: SINTER,
  SINTERCARD: SINTERCARD,
  sInterCard: SINTERCARD,
  SINTERSTORE: SINTERSTORE,
  sInterStore: SINTERSTORE,
  SET: SET,
  set: SET,
  SETBIT: SETBIT,
  setBit: SETBIT,
  SETEX: SETEX,
  setEx: SETEX,
  SETNX: SETNX,
  setNX: SETNX,
  SETRANGE: SETRANGE,
  setRange: SETRANGE,
  SISMEMBER: SISMEMBER,
  sIsMember: SISMEMBER,
  SMEMBERS: SMEMBERS,
  sMembers: SMEMBERS,
  SMISMEMBER: SMISMEMBER,
  smIsMember: SMISMEMBER,
  SMOVE: SMOVE,
  sMove: SMOVE,
  SORT_RO: SORT_RO,
  sortRo: SORT_RO,
  SORT_STORE: SORT_STORE,
  sortStore: SORT_STORE,
  SORT: SORT,
  sort: SORT,
  SPOP: SPOP,
  sPop: SPOP,
  SPUBLISH: SPUBLISH,
  sPublish: SPUBLISH,
  SRANDMEMBER_COUNT: SRANDMEMBER_COUNT,
  sRandMemberCount: SRANDMEMBER_COUNT,
  SRANDMEMBER: SRANDMEMBER,
  sRandMember: SRANDMEMBER,
  SREM: SREM,
  sRem: SREM,
  SSCAN: SSCAN,
  sScan: SSCAN,
  STRLEN: STRLEN,
  strLen: STRLEN,
  SUNION: SUNION,
  sUnion: SUNION,
  SUNIONSTORE: SUNIONSTORE,
  sUnionStore: SUNIONSTORE,
  TOUCH: TOUCH,
  touch: TOUCH,
  TTL: TTL,
  ttl: TTL,
  TYPE: TYPE,
  type: TYPE,
  UNLINK: UNLINK,
  unlink: UNLINK,
  WATCH: WATCH,
  watch: WATCH,
  XACK: XACK,
  xAck: XACK,
  XADD: XADD,
  xAdd: XADD,
  XAUTOCLAIM_JUSTID: XAUTOCLAIM_JUSTID,
  xAutoClaimJustId: XAUTOCLAIM_JUSTID,
  XAUTOCLAIM: XAUTOCLAIM,
  xAutoClaim: XAUTOCLAIM,
  XCLAIM: XCLAIM,
  xClaim: XCLAIM,
  XCLAIM_JUSTID: XCLAIM_JUSTID,
  xClaimJustId: XCLAIM_JUSTID,
  XDEL: XDEL,
  xDel: XDEL,
  XGROUP_CREATE: XGROUP_CREATE,
  xGroupCreate: XGROUP_CREATE,
  XGROUP_CREATECONSUMER: XGROUP_CREATECONSUMER,
  xGroupCreateConsumer: XGROUP_CREATECONSUMER,
  XGROUP_DELCONSUMER: XGROUP_DELCONSUMER,
  xGroupDelConsumer: XGROUP_DELCONSUMER,
  XGROUP_DESTROY: XGROUP_DESTROY,
  xGroupDestroy: XGROUP_DESTROY,
  XGROUP_SETID: XGROUP_SETID,
  xGroupSetId: XGROUP_SETID,
  XINFO_CONSUMERS: XINFO_CONSUMERS,
  xInfoConsumers: XINFO_CONSUMERS,
  XINFO_GROUPS: XINFO_GROUPS,
  xInfoGroups: XINFO_GROUPS,
  XINFO_STREAM: XINFO_STREAM,
  xInfoStream: XINFO_STREAM,
  XLEN: XLEN,
  xLen: XLEN,
  XPENDING_RANGE: XPENDING_RANGE,
  xPendingRange: XPENDING_RANGE,
  XPENDING: XPENDING,
  xPending: XPENDING,
  XRANGE: XRANGE,
  xRange: XRANGE,
  XREAD: XREAD,
  xRead: XREAD,
  XREADGROUP: XREADGROUP,
  xReadGroup: XREADGROUP,
  XREVRANGE: XREVRANGE,
  xRevRange: XREVRANGE,
  XSETID: XSETID,
  xSetId: XSETID,
  XTRIM: XTRIM,
  xTrim: XTRIM,
  ZADD: ZADD,
  zAdd: ZADD,
  ZCARD: ZCARD,
  zCard: ZCARD,
  ZCOUNT: ZCOUNT,
  zCount: ZCOUNT,
  ZDIFF_WITHSCORES: ZDIFF_WITHSCORES,
  zDiffWithScores: ZDIFF_WITHSCORES,
  ZDIFF: ZDIFF,
  zDiff: ZDIFF,
  ZDIFFSTORE: ZDIFFSTORE,
  zDiffStore: ZDIFFSTORE,
  ZINCRBY: ZINCRBY,
  zIncrBy: ZINCRBY,
  ZINTER_WITHSCORES: ZINTER_WITHSCORES,
  zInterWithScores: ZINTER_WITHSCORES,
  ZINTER: ZINTER,
  zInter: ZINTER,
  ZINTERCARD: ZINTERCARD,
  zInterCard: ZINTERCARD,
  ZINTERSTORE: ZINTERSTORE,
  zInterStore: ZINTERSTORE,
  ZLEXCOUNT: ZLEXCOUNT,
  zLexCount: ZLEXCOUNT,
  ZMPOP: ZMPOP,
  zmPop: ZMPOP,
  ZMSCORE: ZMSCORE,
  zmScore: ZMSCORE,
  ZPOPMAX_COUNT: ZPOPMAX_COUNT,
  zPopMaxCount: ZPOPMAX_COUNT,
  ZPOPMAX: ZPOPMAX,
  zPopMax: ZPOPMAX,
  ZPOPMIN_COUNT: ZPOPMIN_COUNT,
  zPopMinCount: ZPOPMIN_COUNT,
  ZPOPMIN: ZPOPMIN,
  zPopMin: ZPOPMIN,
  ZRANDMEMBER_COUNT_WITHSCORES: ZRANDMEMBER_COUNT_WITHSCORES,
  zRandMemberCountWithScores: ZRANDMEMBER_COUNT_WITHSCORES,
  ZRANDMEMBER_COUNT: ZRANDMEMBER_COUNT,
  zRandMemberCount: ZRANDMEMBER_COUNT,
  ZRANDMEMBER: ZRANDMEMBER,
  zRandMember: ZRANDMEMBER,
  ZRANGE_WITHSCORES: ZRANGE_WITHSCORES,
  zRangeWithScores: ZRANGE_WITHSCORES,
  ZRANGE: ZRANGE,
  zRange: ZRANGE,
  ZRANGEBYLEX: ZRANGEBYLEX,
  zRangeByLex: ZRANGEBYLEX,
  ZRANGEBYSCORE_WITHSCORES: ZRANGEBYSCORE_WITHSCORES,
  zRangeByScoreWithScores: ZRANGEBYSCORE_WITHSCORES,
  ZRANGEBYSCORE: ZRANGEBYSCORE,
  zRangeByScore: ZRANGEBYSCORE,
  ZRANGESTORE: ZRANGESTORE,
  zRangeStore: ZRANGESTORE,
  ZRANK: ZRANK,
  zRank: ZRANK,
  ZREM: ZREM,
  zRem: ZREM,
  ZREMRANGEBYLEX: ZREMRANGEBYLEX,
  zRemRangeByLex: ZREMRANGEBYLEX,
  ZREMRANGEBYRANK: ZREMRANGEBYRANK,
  zRemRangeByRank: ZREMRANGEBYRANK,
  ZREMRANGEBYSCORE: ZREMRANGEBYSCORE,
  zRemRangeByScore: ZREMRANGEBYSCORE,
  ZREVRANK: ZREVRANK,
  zRevRank: ZREVRANK,
  ZSCAN: ZSCAN,
  zScan: ZSCAN,
  ZSCORE: ZSCORE,
  zScore: ZSCORE,
  ZUNION_WITHSCORES: ZUNION_WITHSCORES,
  zUnionWithScores: ZUNION_WITHSCORES,
  ZUNION: ZUNION,
  zUnion: ZUNION,
  ZUNIONSTORE: ZUNIONSTORE,
  zUnionStore: ZUNIONSTORE
};
