// @faker-js/faker/dist/cjs/modules/image/providers/lorempicsum.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var u = Object.defineProperty;
var t = Object.getOwnPropertyDescriptor;
var l = Object.getOwnPropertyNames;
var c = Object.prototype.hasOwnProperty;
var p = function p(m, e) {
    for (var r in e) u(m, r, {
      get: e[r],
      enumerable: !0
    });
  },
  d = function d(m, e, r, a) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(l(e)),
        _step;
      try {
        var _loop = function _loop() {
          var i = _step.value;
          !c.call(m, i) && i !== r && u(m, i, {
            get: function get() {
              return e[i];
            },
            enumerable: !(a = t(e, i)) || a.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return m;
  };
var g = function g(m) {
  return d(u({}, "__esModule", {
    value: !0
  }), m);
};
var b = {};
p(b, {
  LoremPicsum: function LoremPicsum() {
    return f;
  }
});
module.exports = g(b);
var s = require("../../../internal/deprecated");
var f = /*#__PURE__*/function () {
  function f(e) {
    _classCallCheck(this, f);
    this.faker = e;
  }
  _createClass(f, [{
    key: "image",
    value: function image(e, r, a, i) {
      return (0, s.deprecated)({
        deprecated: "faker.lorempicsum.image",
        proposed: "faker.image.urlPicsumPhotos",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, a, i);
    }
  }, {
    key: "imageGrayscale",
    value: function imageGrayscale(e, r, a) {
      return (0, s.deprecated)({
        deprecated: "faker.lorempicsum.imageGrayscale",
        proposed: "faker.image.urlPicsumPhotos",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, a);
    }
  }, {
    key: "imageBlurred",
    value: function imageBlurred(e, r, a) {
      return (0, s.deprecated)({
        deprecated: "faker.lorempicsum.imageBlurred",
        proposed: "faker.image.urlPicsumPhotos",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, void 0, a);
    }
  }, {
    key: "imageRandomSeeded",
    value: function imageRandomSeeded(e, r, a, i, n) {
      return (0, s.deprecated)({
        deprecated: "faker.lorempicsum.imageRandomSeeded",
        proposed: "faker.image.urlPicsumPhotos",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, a, i, n);
    }
  }, {
    key: "imageUrl",
    value: function imageUrl(e, r, a, i, n) {
      (0, s.deprecated)({
        deprecated: "faker.lorempicsum.imageUrl",
        proposed: "faker.image.urlPicsumPhotos",
        since: "8.0",
        until: "9.0"
      }), e = e || 640, r = r || 480;
      var o = "https://picsum.photos";
      return n && (o += "/seed/".concat(n)), o += "/".concat(e, "/").concat(r), a && i ? "".concat(o, "?grayscale&blur=").concat(i) : a ? "".concat(o, "?grayscale") : i ? "".concat(o, "?blur=").concat(i) : o;
    }
  }]);
  return f;
}();
0 && (module.exports = {
  LoremPicsum: LoremPicsum
});
