// @faker-js/faker/dist/cjs/modules/lorem/index.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var s = Object.defineProperty;
var o = Object.getOwnPropertyDescriptor;
var u = Object.getOwnPropertyNames;
var h = Object.prototype.hasOwnProperty;
var l = function l(n, e) {
    for (var r in e) s(n, r, {
      get: e[r],
      enumerable: !0
    });
  },
  p = function p(n, e, r, i) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(u(e)),
        _step;
      try {
        var _loop = function _loop() {
          var t = _step.value;
          !h.call(n, t) && t !== r && s(n, t, {
            get: function get() {
              return e[t];
            },
            enumerable: !(i = o(e, t)) || i.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return n;
  };
var b = function b(n) {
  return p(s({}, "__esModule", {
    value: !0
  }), n);
};
var c = {};
l(c, {
  LoremModule: function LoremModule() {
    return m;
  }
});
module.exports = b(c);
var a = require("../word/filterWordListByLength");
var m = /*#__PURE__*/function () {
  function m(e) {
    _classCallCheck(this, m);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(m.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var r = _step2.value;
        r === "constructor" || typeof this[r] != "function" || (this[r] = this[r].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
  }
  _createClass(m, [{
    key: "word",
    value: function word() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var r = typeof e == "number" ? {
        length: e
      } : e;
      return this.faker.helpers.arrayElement((0, a.filterWordListByLength)(_objectSpread(_objectSpread({}, r), {}, {
        wordList: this.faker.definitions.lorem.words
      })));
    }
  }, {
    key: "words",
    value: function words() {
      var _this = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
      return this.faker.helpers.multiple(function () {
        return _this.word();
      }, {
        count: e
      }).join(" ");
    }
  }, {
    key: "sentence",
    value: function sentence() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        min: 3,
        max: 10
      };
      var r = this.words(e);
      return "".concat(r.charAt(0).toUpperCase() + r.substring(1), ".");
    }
  }, {
    key: "slug",
    value: function slug() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
      var r = this.words(e);
      return this.faker.helpers.slugify(r);
    }
  }, {
    key: "sentences",
    value: function sentences() {
      var _this2 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        min: 2,
        max: 6
      };
      var r = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : " ";
      return this.faker.helpers.multiple(function () {
        return _this2.sentence();
      }, {
        count: e
      }).join(r);
    }
  }, {
    key: "paragraph",
    value: function paragraph() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
      return this.sentences(e);
    }
  }, {
    key: "paragraphs",
    value: function paragraphs() {
      var _this3 = this;
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
      var r = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "\n";
      return this.faker.helpers.multiple(function () {
        return _this3.paragraph();
      }, {
        count: e
      }).join(r);
    }
  }, {
    key: "text",
    value: function text() {
      var e = ["sentence", "sentences", "paragraph", "paragraphs", "lines"],
        r = this.faker.helpers.arrayElement(e);
      return "".concat(this[r]());
    }
  }, {
    key: "lines",
    value: function lines() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        min: 1,
        max: 5
      };
      return this.sentences(e, "\n");
    }
  }]);
  return m;
}();
0 && (module.exports = {
  LoremModule: LoremModule
});
