// domutils/lib/esm/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  isTag: true,
  isCDATA: true,
  isText: true,
  isComment: true,
  isDocument: true,
  hasChildren: true
};
Object.defineProperty(exports, "hasChildren", {
  enumerable: true,
  get: function get() {
    return _domhandler.hasChildren;
  }
});
Object.defineProperty(exports, "isCDATA", {
  enumerable: true,
  get: function get() {
    return _domhandler.isCDATA;
  }
});
Object.defineProperty(exports, "isComment", {
  enumerable: true,
  get: function get() {
    return _domhandler.isComment;
  }
});
Object.defineProperty(exports, "isDocument", {
  enumerable: true,
  get: function get() {
    return _domhandler.isDocument;
  }
});
Object.defineProperty(exports, "isTag", {
  enumerable: true,
  get: function get() {
    return _domhandler.isTag;
  }
});
Object.defineProperty(exports, "isText", {
  enumerable: true,
  get: function get() {
    return _domhandler.isText;
  }
});
var _stringify = require("./stringify.js");
Object.keys(_stringify).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _stringify[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _stringify[key];
    }
  });
});
var _traversal = require("./traversal.js");
Object.keys(_traversal).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _traversal[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _traversal[key];
    }
  });
});
var _manipulation = require("./manipulation.js");
Object.keys(_manipulation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _manipulation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _manipulation[key];
    }
  });
});
var _querying = require("./querying.js");
Object.keys(_querying).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _querying[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _querying[key];
    }
  });
});
var _legacy = require("./legacy.js");
Object.keys(_legacy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _legacy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _legacy[key];
    }
  });
});
var _helpers = require("./helpers.js");
Object.keys(_helpers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _helpers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _helpers[key];
    }
  });
});
var _feeds = require("./feeds.js");
Object.keys(_feeds).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _feeds[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _feeds[key];
    }
  });
});
var _domhandler = require("domhandler");
