// @faker-js/faker/dist/cjs/modules/image/index.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
var u = Object.defineProperty;
var f = Object.getOwnPropertyDescriptor;
var d = Object.getOwnPropertyNames;
var b = Object.prototype.hasOwnProperty;
var k = function k(l, e) {
    for (var r in e) u(l, r, {
      get: e[r],
      enumerable: !0
    });
  },
  y = function y(l, e, r, t) {
    if (e && _typeof(e) == "object" || typeof e == "function") {
      var _iterator = _createForOfIteratorHelper(d(e)),
        _step;
      try {
        var _loop = function _loop() {
          var a = _step.value;
          !b.call(l, a) && a !== r && u(l, a, {
            get: function get() {
              return e[a];
            },
            enumerable: !(t = f(e, a)) || t.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return l;
  };
var $ = function $(l) {
  return y(u({}, "__esModule", {
    value: !0
  }), l);
};
var w = {};
k(w, {
  ImageModule: function ImageModule() {
    return g;
  }
});
module.exports = $(w);
var i = require("../../internal/deprecated"),
  c = require("./providers/lorempicsum"),
  h = require("./providers/placeholder"),
  p = require("./providers/unsplash");
var g = /*#__PURE__*/function () {
  function g(e) {
    _classCallCheck(this, g);
    this.faker = e;
    var _iterator2 = _createForOfIteratorHelper(Object.getOwnPropertyNames(g.prototype)),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var r = _step2.value;
        r === "constructor" || typeof this[r] != "function" || (this[r] = this[r].bind(this));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
    this.unsplash = new p.Unsplash(this.faker), this.lorempicsum = new c.LoremPicsum(this.faker), this.placeholder = new h.Placeholder(this.faker);
  }
  _createClass(g, [{
    key: "avatar",
    value: function avatar() {
      return this.faker.helpers.arrayElement([this.avatarLegacy, this.avatarGitHub])();
    }
  }, {
    key: "avatarGitHub",
    value: function avatarGitHub() {
      return "https://avatars.githubusercontent.com/u/".concat(this.faker.number["int"](1e8));
    }
  }, {
    key: "avatarLegacy",
    value: function avatarLegacy() {
      return "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/".concat(this.faker.number["int"](1249), ".jpg");
    }
  }, {
    key: "url",
    value: function url() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$width = e.width,
        r = _e$width === void 0 ? 640 : _e$width,
        _e$height = e.height,
        t = _e$height === void 0 ? 480 : _e$height;
      return this.faker.helpers.arrayElement([this.urlLoremFlickr, this.urlPicsumPhotos])({
        width: r,
        height: t
      });
    }
  }, {
    key: "urlLoremFlickr",
    value: function urlLoremFlickr() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$width2 = e.width,
        r = _e$width2 === void 0 ? 640 : _e$width2,
        _e$height2 = e.height,
        t = _e$height2 === void 0 ? 480 : _e$height2,
        a = e.category;
      return "https://loremflickr.com/".concat(r, "/").concat(t).concat(a != null ? "/".concat(a) : "", "?lock=").concat(this.faker.number["int"]());
    }
  }, {
    key: "urlPicsumPhotos",
    value: function urlPicsumPhotos() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$width3 = e.width,
        r = _e$width3 === void 0 ? 640 : _e$width3,
        _e$height3 = e.height,
        t = _e$height3 === void 0 ? 480 : _e$height3,
        _e$grayscale = e.grayscale,
        a = _e$grayscale === void 0 ? !1 : _e$grayscale,
        n = e.blur;
      var s = "https://picsum.photos/seed/".concat(this.faker.string.alphanumeric({
        length: {
          min: 5,
          max: 10
        }
      }), "/").concat(r, "/").concat(t);
      var m = a === !0,
        o = typeof n == "number" && n >= 1 && n <= 10;
      return (m || o) && (s += "?", m && (s += "grayscale"), m && o && (s += "&"), o && (s += "blur=".concat(n))), s;
    }
  }, {
    key: "urlPlaceholder",
    value: function urlPlaceholder() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$width4 = e.width,
        r = _e$width4 === void 0 ? this.faker.number["int"]({
          min: 1,
          max: 3999
        }) : _e$width4,
        _e$height4 = e.height,
        t = _e$height4 === void 0 ? this.faker.number["int"]({
          min: 1,
          max: 3999
        }) : _e$height4,
        _e$backgroundColor = e.backgroundColor,
        a = _e$backgroundColor === void 0 ? this.faker.color.rgb({
          format: "hex",
          prefix: ""
        }) : _e$backgroundColor,
        _e$textColor = e.textColor,
        n = _e$textColor === void 0 ? this.faker.color.rgb({
          format: "hex",
          prefix: ""
        }) : _e$textColor,
        _e$format = e.format,
        s = _e$format === void 0 ? this.faker.helpers.arrayElement(["gif", "jpeg", "jpg", "png", "webp"]) : _e$format,
        _e$text = e.text,
        m = _e$text === void 0 ? this.faker.lorem.words() : _e$text;
      var o = "https://via.placeholder.com";
      return o += "/".concat(r), o += "x".concat(t), o += "/".concat(a), o += "/".concat(n), o += ".".concat(s), o += "?text=".concat(encodeURIComponent(m)), o;
    }
  }, {
    key: "dataUri",
    value: function dataUri() {
      var e = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      var _e$width5 = e.width,
        r = _e$width5 === void 0 ? 640 : _e$width5,
        _e$height5 = e.height,
        t = _e$height5 === void 0 ? 480 : _e$height5,
        _e$color = e.color,
        a = _e$color === void 0 ? "grey" : _e$color,
        n = "<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" baseProfile=\"full\" width=\"".concat(r, "\" height=\"").concat(t, "\"><rect width=\"100%\" height=\"100%\" fill=\"").concat(a, "\"/><text x=\"").concat(r / 2, "\" y=\"").concat(t / 2, "\" font-size=\"20\" alignment-baseline=\"middle\" text-anchor=\"middle\" fill=\"white\">").concat(r, "x").concat(t, "</text></svg>");
      return "data:image/svg+xml;charset=UTF-8," + encodeURIComponent(n);
    }
  }, {
    key: "image",
    value: function image(e, r, t) {
      (0, i.deprecated)({
        deprecated: "faker.image.image",
        proposed: "faker.image.url",
        since: "8.0",
        until: "9.0"
      });
      var a = ["abstract", "animals", "business", "cats", "city", "food", "nightlife", "fashion", "people", "nature", "sports", "technics", "transport"];
      return this[this.faker.helpers.arrayElement(a)](e, r, t);
    }
  }, {
    key: "imageUrl",
    value: function imageUrl(e, r, t, a) {
      (0, i.deprecated)({
        deprecated: "faker.image.imageUrl",
        proposed: "faker.image.url",
        since: "8.0",
        until: "9.0"
      }), e = e || 640, r = r || 480;
      var n = "https://loremflickr.com/".concat(e, "/").concat(r);
      return t != null && (n += "/".concat(t)), a && (n += "?lock=".concat(this.faker.number["int"]())), n;
    }
  }, {
    key: "abstract",
    value: function abstract(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.abstract",
        proposed: "faker.image.urlLoremFlickr({ category: 'abstract' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "abstract", t);
    }
  }, {
    key: "animals",
    value: function animals(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.animals",
        proposed: "faker.image.urlLoremFlickr({ category: 'animals' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "animals", t);
    }
  }, {
    key: "business",
    value: function business(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.business",
        proposed: "faker.image.urlLoremFlickr({ category: 'business' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "business", t);
    }
  }, {
    key: "cats",
    value: function cats(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.cats",
        proposed: "faker.image.urlLoremFlickr({ category: 'cats' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "cats", t);
    }
  }, {
    key: "city",
    value: function city(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.city",
        proposed: "faker.image.urlLoremFlickr({ category: 'city' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "city", t);
    }
  }, {
    key: "food",
    value: function food(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.food",
        proposed: "faker.image.urlLoremFlickr({ category: 'food' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "food", t);
    }
  }, {
    key: "nightlife",
    value: function nightlife(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.nightlife",
        proposed: "faker.image.urlLoremFlickr({ category: 'nightlife' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "nightlife", t);
    }
  }, {
    key: "fashion",
    value: function fashion(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.fashion",
        proposed: "faker.image.urlLoremFlickr({ category: 'fashion' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "fashion", t);
    }
  }, {
    key: "people",
    value: function people(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.people",
        proposed: "faker.image.urlLoremFlickr({ category: 'people' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "people", t);
    }
  }, {
    key: "nature",
    value: function nature(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.nature",
        proposed: "faker.image.urlLoremFlickr({ category: 'nature' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "nature", t);
    }
  }, {
    key: "sports",
    value: function sports(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.sports",
        proposed: "faker.image.urlLoremFlickr({ category: 'sports' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "sports", t);
    }
  }, {
    key: "technics",
    value: function technics(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.technics",
        proposed: "faker.image.urlLoremFlickr({ category: 'technics' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "technics", t);
    }
  }, {
    key: "transport",
    value: function transport(e, r, t) {
      return (0, i.deprecated)({
        deprecated: "faker.image.transport",
        proposed: "faker.image.urlLoremFlickr({ category: 'transport' }) or faker.image.url",
        since: "8.0",
        until: "9.0"
      }), this.imageUrl(e, r, "transport", t);
    }
  }]);
  return g;
}();
0 && (module.exports = {
  ImageModule: ImageModule
});
